<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Solutions • rpharma2020tidytranscriptomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Solutions">
<meta property="og:description" content="rpharma2020tidytranscriptomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rpharma2020tidytranscriptomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/background.html">Background</a>
</li>
<li>
  <a href="../articles/tidytranscriptomics.html">Workshop</a>
</li>
<li>
  <a href="../articles/solutions.html">Solutions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/rpharma2020_tidytranscriptomics">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://hub.docker.com/repository/docker/stemangiola/rpharma2020_tidytranscriptomics">
    <span class="fa fa-docker"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="solutions_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Solutions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/rpharma2020_tidytranscriptomics/blob/master/vignettes/solutions.Rmd"><code>vignettes/solutions.Rmd</code></a></small>
      <div class="hidden name"><code>solutions.Rmd</code></div>

    </div>

    
    
<p>Suggested answers to the workshop questions are below. You might have some different code e.g. to customise the volcano plot as you like. Feel free to comment on any of these solutions in the workshop website as described <a href="https://github.com/stemangiola/rpharma2020_tidytranscriptomics/blob/master/CONTRIBUTING.md">here</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co"># load libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>

<span class="co"># tidyverse core packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://stringr.tidyverse.org">stringr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>

<span class="co"># tidyverse-friendly packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com">plotly</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/slowkow/ggrepel">ggrepel</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.r-project.org">tidyHeatmap</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidybulk</span><span class="op">)</span></pre></div>
<div id="part-1-bulk-rna-seq-core" class="section level1">
<h1 class="hasAnchor">
<a href="#part-1-bulk-rna-seq-core" class="anchor"></a>Part 1 Bulk RNA-seq Core</h1>
<div id="airway-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#airway-dataset" class="anchor"></a>Airway dataset</h2>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># Set up data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"airway"</span><span class="op">)</span>

<span class="va">counts_tt</span> <span class="op">&lt;-</span> 
  <span class="va">airway</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<div id="what-fraction-of-variance-is-explained-by-pc3" class="section level3">
<h3 class="hasAnchor">
<a href="#what-fraction-of-variance-is-explained-by-pc3" class="anchor"></a>What fraction of variance is explained by PC3?</h3>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">counts_tt</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"PCA"</span>, .dims<span class="op">=</span><span class="fl">3</span><span class="op">)</span></pre></div>
<pre><code>## Warning in eval(dots[[i]][[action]], env, env): tidybulk says: highly abundant
## transcripts were not identified (i.e. identify_abundant()) or filtered (i.e.,
## keep_abundant), therefore this operation will be performed on unfiltered
## data. In rare occasions this could be wanted. In standard whole-transcriptome
## workflows is generally unwanted.

## Warning in eval(dots[[i]][[action]], env, env): tidybulk says: highly abundant
## transcripts were not identified (i.e. identify_abundant()) or filtered (i.e.,
## keep_abundant), therefore this operation will be performed on unfiltered
## data. In rare occasions this could be wanted. In standard whole-transcriptome
## workflows is generally unwanted.</code></pre>
<pre><code>## Getting the 500 most variable genes</code></pre>
<pre><code>## Fraction of variance explained by the selected principal components</code></pre>
<pre><code>## # A tibble: 3 x 2
##   `Fraction of variance`    PC
##                    &lt;dbl&gt; &lt;int&gt;
## 1                  0.314     1
## 2                  0.300     2
## 3                  0.177     3</code></pre>
<pre><code>## tidybulk says: to access the raw results do `attr(..., "internals")$PCA`</code></pre>
<pre><code>## # A tibble: 512,816 x 18
##    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment
##    &lt;fct&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     
##  1 ENSG00… SRR10…    679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
##  2 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
##  3 ENSG00… SRR10…    467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
##  4 ENSG00… SRR10…    260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
##  5 ENSG00… SRR10…     60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
##  6 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
##  7 ENSG00… SRR10…   3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
##  8 ENSG00… SRR10…   1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
##  9 ENSG00… SRR10…    519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
## 10 ENSG00… SRR10…    394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
## # … with 512,806 more rows, and 8 more variables: Sample &lt;fct&gt;,
## #   BioSample &lt;fct&gt;, TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, counts_scaled &lt;dbl&gt;,
## #   PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;</code></pre>
</div>
<div id="how-many-differentially-expressed-transcripts-are-there-for-fdr-0-05-if-we-did-not-include-cell-line-in-the-formula" class="section level3">
<h3 class="hasAnchor">
<a href="#how-many-differentially-expressed-transcripts-are-there-for-fdr-0-05-if-we-did-not-include-cell-line-in-the-formula" class="anchor"></a>How many differentially expressed transcripts are there for FDR &lt; 0.05 if we did not include cell line in the formula?</h3>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="va">counts_tt</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span><span class="op">(</span>factor_of_interest<span class="op">=</span><span class="va">dex</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
      .formula <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dex</span>,
      .contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"dextrt - dexuntrt"</span><span class="op">)</span>,
      omit_contrast_in_colnames <span class="op">=</span> <span class="cn">TRUE</span>
    <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>num_de <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html">n_distinct</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span></pre></div>
<pre><code>## tidybulk says: All methods use raw counts, 
## irrespective of if scale_abundance or adjust_abundance have been calculated, 
## therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</code></pre>
<pre><code>## tidybulk says: The design column names are "dextrt, dexuntrt"</code></pre>
<pre><code>## tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</code></pre>
<pre><code>## # A tibble: 1 x 1
##   num_de
##    &lt;int&gt;
## 1   1952</code></pre>
</div>
</div>
<div id="pasilla-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#pasilla-dataset" class="anchor"></a>Pasilla dataset</h2>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="co"># load data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"pasilla"</span>, package <span class="op">=</span> <span class="st">"rpharma2020tidytranscriptomics"</span><span class="op">)</span>

<span class="co"># create tidybulk tibble</span>
<span class="va">counts_tt</span> <span class="op">&lt;-</span>
    <span class="va">pasilla</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">mapIds</a></span><span class="op">(</span><span class="fu">org.Dm.eg.db</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/org.Dm.eg.db/man/org.Dm.egBASE.html">org.Dm.eg.db</a></span>, keys<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span>, keytype <span class="op">=</span> <span class="st">"FLYBASE"</span>, column<span class="op">=</span><span class="st">"SYMBOL"</span>, multiVals <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span><span class="op">)</span></pre></div>
<pre><code>## </code></pre>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="co"># filter counts</span>
<span class="va">counts_filtered</span> <span class="op">&lt;-</span> <span class="va">counts_tt</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">condition</span><span class="op">)</span>

<span class="co"># scale counts</span>
<span class="va">counts_scaled</span> <span class="op">&lt;-</span> <span class="va">counts_filtered</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<div id="what-is-the-fraction-of-variance-for-pc1" class="section level3">
<h3 class="hasAnchor">
<a href="#what-is-the-fraction-of-variance-for-pc1" class="anchor"></a>What is the Fraction of Variance for PC1?</h3>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="va">counts_scal_PCA</span> <span class="op">&lt;-</span>
  <span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></pre></div>
<pre><code>## Getting the 500 most variable genes</code></pre>
<pre><code>## Fraction of variance explained by the selected principal components</code></pre>
<pre><code>## # A tibble: 2 x 2
##   `Fraction of variance`    PC
##                    &lt;dbl&gt; &lt;int&gt;
## 1                  0.471     1
## 2                  0.250     2</code></pre>
<pre><code>## tidybulk says: to access the raw results do `attr(..., "internals")$PCA`</code></pre>
</div>
<div id="how-many-differentially-expressed-genes-are-there-for-treated-vs-untreated-fdr-0-05" class="section level3">
<h3 class="hasAnchor">
<a href="#how-many-differentially-expressed-genes-are-there-for-treated-vs-untreated-fdr-0-05" class="anchor"></a>How many differentially expressed genes are there for treated vs untreated (FDR &lt; 0.05)?</h3>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="va">counts_de</span> <span class="op">&lt;-</span>
  <span class="va">counts_tt</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>.formula <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">condition</span> <span class="op">+</span> <span class="va">type</span>, 
                              .contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"conditiontreated - conditionuntreated"</span><span class="op">)</span>, 
                              omit_contrast_in_colnames <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
<pre><code>## tidybulk says: All methods use raw counts, 
## irrespective of if scale_abundance or adjust_abundance have been calculated, 
## therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</code></pre>
<pre><code>## Warning in eval(dots[[i]][[action]], env, env): tidybulk says: highly abundant
## transcripts were not identified (i.e. identify_abundant()) or filtered (i.e.,
## keep_abundant), therefore this operation will be performed on unfiltered
## data. In rare occasions this could be wanted. In standard whole-transcriptome
## workflows is generally unwanted.</code></pre>
<pre><code>## tidybulk says: The design column names are "conditiontreated, conditionuntreated, typesingle_end"</code></pre>
<pre><code>## tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="va">counts_de</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>num_de <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html">n_distinct</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span></pre></div>
<pre><code>## # A tibble: 1 x 1
##   num_de
##    &lt;int&gt;
## 1   1471</code></pre>
</div>
</div>
<div id="what-is-the-fbgn-id-of-the-10th-most-differentially-expressed-gene-by-smallest-p-value" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-the-fbgn-id-of-the-10th-most-differentially-expressed-gene-by-smallest-p-value" class="anchor"></a>What is the FBgn id of the 10th most differentially expressed gene (by smallest P value)?</h2>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="va">topgenes</span> <span class="op">&lt;-</span> <span class="va">counts_de</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">PValue</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>

<span class="va">topgenes</span></pre></div>
<pre><code>## # A tibble: 10 x 7
##    feature     symbol logFC logCPM     F    PValue       FDR
##    &lt;chr&gt;       &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
##  1 FBgn0039155 Kal1   -4.61   5.88 1119. 4.58e-243 6.69e-239
##  2 FBgn0025111 Ant2    2.87   6.92  794. 1.61e-173 1.18e-169
##  3 FBgn0003360 sesB   -3.11   8.45  740. 7.01e-162 3.41e-158
##  4 FBgn0039827 CG1544 -4.16   4.39  599. 1.08e-131 3.96e-128
##  5 FBgn0035085 CG3770 -2.57   5.68  473. 1.46e-104 4.28e-101
##  6 FBgn0034736 gas    -3.52   4.18  449. 3.53e- 99 8.59e- 96
##  7 FBgn0000071 Ama     2.63   4.79  439. 4.70e- 97 9.81e- 94
##  8 FBgn0026562 SPARC  -2.46  11.8   432. 1.57e- 95 2.86e- 92
##  9 FBgn0029167 Hml    -2.18   8.22  415. 5.22e- 92 8.47e- 89
## 10 FBgn0035189 CG9119  2.89   4.12  402. 4.65e- 89 6.79e- 86</code></pre>
</div>
<div id="extra" class="section level2">
<h2 class="hasAnchor">
<a href="#extra" class="anchor"></a>Extra</h2>
<div id="question-1-4" class="section level3">
<h3 class="hasAnchor">
<a href="#question-1-4" class="anchor"></a>Question 1.4</h3>
<p>What code can generate a heatmap of variable genes (starting from count_scaled)?</p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="va">counts_scaled</span> <span class="op">%&gt;%</span> 
    
    <span class="co"># extract 500 most variable genes</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_variable-methods.html">keep_variable</a></span><span class="op">(</span> .abundance <span class="op">=</span> <span class="va">counts_scaled</span>, top <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">%&gt;%</span>
    
    <span class="co"># create heatmap</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/heatmap-methods.html">heatmap</a></span><span class="op">(</span>
          .column <span class="op">=</span> <span class="va">sample</span>,
          .row <span class="op">=</span> <span class="va">feature</span>,
          .value <span class="op">=</span> <span class="va">counts_scaled</span>,
          annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">condition</span>, <span class="va">type</span><span class="op">)</span>,
          transform <span class="op">=</span> <span class="va">log1p</span> 
      <span class="op">)</span></pre></div>
<pre><code>## Getting the 500 most variable genes</code></pre>
<pre><code>## Warning: The `annotation` argument of `heatmap()` is deprecated as of tidyHeatmap 1.1.0.
## Please use the new annotation framework instead: heatmap(...) %&gt;% add_tile(...) %&gt;% add_point(...) %&gt;% add_bar() %&gt;% add_line() %&gt;% ...
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_warnings()` to see where this warning was generated.</code></pre>
<p><img src="solutions_files/figure-html/unnamed-chunk-9-1.png" width="70%"></p>
</div>
<div id="question-1-5" class="section level3">
<h3 class="hasAnchor">
<a href="#question-1-5" class="anchor"></a>Question 1.5</h3>
<p>What code can you use to visualise expression of the pasilla gene (gene id: FBgn0261552)</p>
<div class="sourceCode" id="cb36"><pre class="downlit">
<span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
    
    <span class="co"># extract counts for pasilla gene</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">==</span> <span class="st">"FBgn0261552"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    
    <span class="co"># make stripchart</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">condition</span>, y <span class="op">=</span> <span class="va">counts_scaled</span> <span class="op">+</span> <span class="fl">1</span>, fill <span class="op">=</span><span class="va">condition</span>, label <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p><img src="solutions_files/figure-html/unnamed-chunk-10-1.png" width="70%"></p>
</div>
<div id="question-1-6" class="section level3">
<h3 class="hasAnchor">
<a href="#question-1-6" class="anchor"></a>Question 1.6</h3>
<p>What code can generate an interactive volcano plot that has gene ids showing on hover?</p>
<div class="sourceCode" id="cb37"><pre class="downlit">
<span class="va">p</span> <span class="op">&lt;-</span> <span class="va">counts_de</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Subset data</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>significant <span class="op">=</span> <span class="va">FDR</span><span class="op">&lt;</span><span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;=</span><span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Plot</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span>, label<span class="op">=</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">significant</span>, size <span class="op">=</span> <span class="va">significant</span>, alpha<span class="op">=</span><span class="va">significant</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    
    <span class="co"># Custom scales</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"log10_reverse"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"#e11f28"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/plotly/man/ggplotly.html">ggplotly</a></span><span class="op">(</span><span class="va">p</span>, tooltip <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"text"</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>Tip: You can use “text” instead of “label” if you don’t want the column name to show up in the hover e.g. above will give “FBgn0261552” rather than “feature:FBgn0261552”.</p>
</div>
<div id="question-1-7" class="section level3">
<h3 class="hasAnchor">
<a href="#question-1-7" class="anchor"></a>Question 1.7</h3>
<p>What code can generate a heatmap of the top 100 DE genes?</p>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="va">top100</span> <span class="op">&lt;-</span> 
    <span class="va">counts_de</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">PValue</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>

<span class="va">counts_scaled</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">%in%</span> <span class="va">top100</span><span class="op">$</span><span class="va">feature</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/heatmap-methods.html">heatmap</a></span><span class="op">(</span>
          .column <span class="op">=</span> <span class="va">sample</span>,
          .row <span class="op">=</span> <span class="va">feature</span>,
          .value <span class="op">=</span> <span class="va">counts_scaled</span>,
          annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">condition</span>, <span class="va">type</span><span class="op">)</span>,
          transform <span class="op">=</span> <span class="va">log1p</span> 
      <span class="op">)</span></pre></div>
<p><img src="solutions_files/figure-html/unnamed-chunk-12-1.png" width="70%"></p>
</div>
</div>
</div>
<div id="part-2-bulk-rna-seq-extended" class="section level1">
<h1 class="hasAnchor">
<a href="#part-2-bulk-rna-seq-extended" class="anchor"></a>Part 2 Bulk RNA-seq Extended</h1>
<div id="comparison-of-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#comparison-of-methods" class="anchor"></a>Comparison of methods</h2>
<div id="which-method-detects-the-most-differentially-abundant-transcripts-p-value-adjusted-for-multiple-testing-0-05-fdr-adj-p-val-padj" class="section level3">
<h3 class="hasAnchor">
<a href="#which-method-detects-the-most-differentially-abundant-transcripts-p-value-adjusted-for-multiple-testing-0-05-fdr-adj-p-val-padj" class="anchor"></a>Which method detects the most differentially abundant transcripts, p value adjusted for multiple testing &lt; 0.05 (FDR, adj.P.Val, padj)?</h3>
<div class="sourceCode" id="cb39"><pre class="downlit">
<span class="co"># Set up data</span>
<span class="va">pasilla_de</span> <span class="op">&lt;-</span> 
  <span class="fu">rpharma2020tidytranscriptomics</span><span class="fu">::</span><span class="va"><a href="../reference/pasilla.html">pasilla</a></span> <span class="op">%&gt;%</span> 
  
  <span class="co"># Convert SE object to tibble</span>
  <span class="va">tidybulk</span> <span class="op">%&gt;%</span>
  
   <span class="co"># Scale abundance for plotting</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest<span class="op">=</span><span class="va">condition</span><span class="op">)</span> 

<span class="va">de_all</span> <span class="op">&lt;-</span> 
  
  <span class="va">pasilla_de</span> <span class="op">%&gt;%</span>
  
  <span class="co"># edgeR QLT</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">condition</span> <span class="op">+</span> <span class="va">type</span>, 
    method <span class="op">=</span> <span class="st">"edger_quasi_likelihood"</span>,
    prefix <span class="op">=</span> <span class="st">"edgerQLT_"</span>
  <span class="op">)</span>  <span class="op">%&gt;%</span>
  
  <span class="co"># edgeR LRT</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">condition</span> <span class="op">+</span> <span class="va">type</span>, 
    method <span class="op">=</span> <span class="st">"edger_likelihood_ratio"</span>,
    prefix <span class="op">=</span> <span class="st">"edgerLR_"</span>
  <span class="op">)</span>  <span class="op">%&gt;%</span>
  
  <span class="co"># limma-voom</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">condition</span> <span class="op">+</span> <span class="va">type</span>, 
    method <span class="op">=</span> <span class="st">"limma_voom"</span>,
    prefix <span class="op">=</span> <span class="st">"voom_"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># DESeq2</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">condition</span> <span class="op">+</span> <span class="va">type</span>, 
    method <span class="op">=</span> <span class="st">"deseq2"</span>,
    prefix <span class="op">=</span> <span class="st">"deseq2_"</span>
  <span class="op">)</span> </pre></div>
<pre><code>## tidybulk says: All methods use raw counts, 
## irrespective of if scale_abundance or adjust_abundance have been calculated, 
## therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</code></pre>
<pre><code>## tidybulk says: The design column names are "(Intercept), conditionuntreated, typesingle_end"</code></pre>
<pre><code>## tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</code></pre>
<pre><code>## tidybulk says: All methods use raw counts, 
## irrespective of if scale_abundance or adjust_abundance have been calculated, 
## therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</code></pre>
<pre><code>## tidybulk says: The design column names are "(Intercept), conditionuntreated, typesingle_end"</code></pre>
<pre><code>## tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</code></pre>
<pre><code>## tidybulk says: All methods use raw counts, 
## irrespective of if scale_abundance or adjust_abundance have been calculated, 
## therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</code></pre>
<pre><code>## tidybulk says: The design column names are "(Intercept), conditionuntreated, typesingle_end"</code></pre>
<pre><code>## tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$voom`</code></pre>
<pre><code>## tidybulk says: All methods use raw counts, 
## irrespective of if scale_abundance or adjust_abundance have been calculated, 
## therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</code></pre>
<pre><code>## Installing acepack needed for analyses</code></pre>
<pre><code>## Installing package into '/tmp/Rtmp3iDSlx/temp_libpath1fe41a89d0b3'
## (as 'lib' is unspecified)</code></pre>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<pre><code>## tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$DESeq2`</code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit">
<span class="va">de_all</span> <span class="op">%&gt;%</span>
    
    <span class="co"># Subset transcript information</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    
    <span class="co"># Reshape for nesting</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>
        cols <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">feature</span>, <span class="va">.abundant</span><span class="op">)</span>,
        names_sep <span class="op">=</span> <span class="st">"_"</span>, 
        names_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"method"</span>, <span class="st">"statistic"</span><span class="op">)</span>, 
        values_to <span class="op">=</span> <span class="st">"value"</span>
    <span class="op">)</span> <span class="op">%&gt;%</span>
    
    <span class="co"># Filter statistic</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">statistic</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FDR"</span>, <span class="st">"adj.P.Val"</span>, <span class="st">"padj"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">value</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span>
    
    <span class="co"># Nesting</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">method</span><span class="op">)</span></pre></div>
<pre><code>## # A tibble: 4 x 2
##   method       n
##   &lt;chr&gt;    &lt;int&gt;
## 1 deseq2    1112
## 2 edgerLR   1383
## 3 edgerQLT  1128
## 4 voom      1133</code></pre>
</div>
</div>
<div id="cell-type-composition" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-composition" class="anchor"></a>Cell type composition</h2>
<div id="what-is-the-most-abundant-cell-type-overall-in-brca-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#what-is-the-most-abundant-cell-type-overall-in-brca-samples" class="anchor"></a>What is the most abundant cell type overall in BRCA samples?</h3>
<div class="sourceCode" id="cb61"><pre class="downlit">
<span class="va">BRCA_cell_type_long</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cell_type</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>m <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">proportion</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Maria Doyle, Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
