<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to Tidy Transcriptomics • rpharma2020tidytranscriptomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to Tidy Transcriptomics">
<meta property="og:description" content="rpharma2020tidytranscriptomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rpharma2020tidytranscriptomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/background.html">Background</a>
</li>
<li>
  <a href="../articles/tidytranscriptomics.html">Workshop</a>
</li>
<li>
  <a href="../articles/solutions.html">Solutions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/rpharma2020_tidytranscriptomics">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://hub.docker.com/repository/docker/stemangiola/rpharma2020_tidytranscriptomics">
    <span class="fa fa-docker"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tidytranscriptomics_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="tidytranscriptomics_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="tidytranscriptomics_files/plotly-binding-4.9.2.1/plotly.js"></script><script src="tidytranscriptomics_files/typedarray-0.1/typedarray.min.js"></script><script src="tidytranscriptomics_files/jquery-1.11.3/jquery.min.js"></script><link href="tidytranscriptomics_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="tidytranscriptomics_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><link href="tidytranscriptomics_files/plotly-htmlwidgets-css-1.52.2/plotly-htmlwidgets.css" rel="stylesheet">
<script src="tidytranscriptomics_files/plotly-main-1.52.2/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to Tidy Transcriptomics</h1>
                        <h4 class="author">Maria Doyle, Peter MacCallum Cancer Centre<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
                        <h4 class="author">Stefano Mangiola, Walter and Eliza Hall Institute<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/rpharma2020_tidytranscriptomics/blob/master/vignettes/tidytranscriptomics.Rmd"><code>vignettes/tidytranscriptomics.Rmd</code></a></small>
      <div class="hidden name"><code>tidytranscriptomics.Rmd</code></div>

    </div>

    
    
<div id="part-1-bulk-rna-seq-core" class="section level1">
<h1 class="hasAnchor">
<a href="#part-1-bulk-rna-seq-core" class="anchor"></a>Part 1 Bulk RNA-seq Core</h1>
<p>This workshop will present how to perform analysis of RNA sequencing data following the tidy data paradigm <span class="citation">(Wickham and others 2014)</span>. The tidy data paradigm provides a standard way to organise data values within a dataset, where each variable is a column, each observation is a row, and data is manipulated using an easy-to-understand vocabulary. Most importantly, the data structure remains consistent across manipulation and analysis functions.</p>
<p>This can be achieved for RNA sequencing data with the <a href="https://stemangiola.github.io/tidybulk/">tidybulk</a>, <a href="https://stemangiola.github.io/tidyHeatmap">tidyHeatmap</a> <span class="citation">(Mangiola and Papenfuss 2020)</span> and <a href="https://www.tidyverse.org/">tidyverse</a> <span class="citation">(Wickham et al. 2019)</span> packages. The tidybulk package provides a tidy data structure and a modular framework for bulk transcriptional analyses. tidyHeatmap provides a tidy implementation of ComplexHeatmap. These packages are part of the tidytranscriptomics suite that introduces a tidy approach to RNA sequencing data representation and analysis</p>
<div id="acknowledgements" class="section level3">
<h3 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h3>
<p>This material was adapted from an R for RNA sequencing workshop first run <a href="http://combine-australia.github.io/2016-05-11-RNAseq/">here</a>. Use of the airway and pasilla datasets was inspired by the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DESeq2 vignette</a>.</p>
<p><img src="../inst/vignettes/tidybulk_logo.png" width="100px"></p>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Measuring gene expression on a genome-wide scale has become common practice over the last two decades or so, with microarrays predominantly used pre-2008. With the advent of next generation sequencing technology in 2008, an increasing number of scientists use this technology to measure and understand changes in gene expression in often complex systems. As sequencing costs have decreased, using RNA sequencing to simultaneously measure the expression of tens of thousands of genes for multiple samples has never been easier. The cost of these experiments has now moved from generating the data to storing and analysing it.</p>
<p>There are many steps involved in analysing an RNA sequencing dataset. Sequenced reads are aligned to a reference genome, then the number of reads mapped to each gene can be counted. This results in a table of counts, which is what we perform statistical analyses on in R. While mapping and counting are important and necessary tasks, today we will be starting from the count data and showing how differential expression analysis can be performed in a friendly way using tidybulk.</p>
<p>First, let’s load all the packages we will need to analyse the data. Note that you should load the <em>tidybulk</em> library after the tidyverse core packages for best integration.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co"># load libraries</span>

<span class="co"># dataset</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>

<span class="co"># annotation</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Dm.eg.db</span><span class="op">)</span>

<span class="co"># tidyverse core packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://readr.tidyverse.org">readr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://stringr.tidyverse.org">stringr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>

<span class="co"># tidyverse-friendly packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.r-project.org">tidyHeatmap</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidybulk</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/slowkow/ggrepel">ggrepel</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com">plotly</a></span><span class="op">)</span></pre></div>
<p>Plot settings. Set the colours and theme we will use for our plots.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># Use colourblind-friendly colours</span>
<span class="va">friendly_cols</span> <span class="op">&lt;-</span> <span class="fu">dittoSeq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html">dittoColors</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Set theme</span>
<span class="va">custom_theme</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
        panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span><span class="op">)</span>,
        panel.grid.major <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>,
        panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,
        text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
        legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        <span class="co">#aspect.ratio = 1,</span>
        strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">10</span>, r <span class="op">=</span> <span class="fl">10</span>, b <span class="op">=</span> <span class="fl">10</span>, l <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
        axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">10</span>, r <span class="op">=</span> <span class="fl">10</span>, b <span class="op">=</span> <span class="fl">10</span>, l <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
        axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">30</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
      <span class="op">)</span>
  <span class="op">)</span></pre></div>
<div id="airway-rna-sequencing-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#airway-rna-sequencing-dataset" class="anchor"></a>Airway RNA sequencing dataset</h3>
<p>Here we will perform our analysis using the data from the <em>airway</em> package. The airway data comes from the paper by <span class="citation">(Himes et al. 2014)</span>; and it includes 8 samples from human airway smooth muscle cells, from 4 cell lines. For each cell line treated (with dexamethasone) and untreated (negative control) a sample has undergone RNA sequencing and gene counts have been generated.</p>
</div>
<div id="setting-up-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#setting-up-the-data" class="anchor"></a>Setting up the data</h3>
<p>The airway data is stored as a <em>RangedSummarizedExperiment</em> object. We can convert a SummarizedExperiment object into a tidybulk tibble, which is the tidyverse table format.</p>
<p>In this workshop we will be using the tidyverse pipe <code><a href="https://rdrr.io/pkg/tidyseurat/man/pipe.html">%&gt;%</a></code>. This ‘pipes’ the output from the command on the left into the command on the right/below. Using the pipe is not essential but it reduces the amount of code we need to write when we have multiple steps (as we’ll see later). It also can make the steps clearer and easier to see. For more details on the pipe see <a href="https://r4ds.had.co.nz/pipes.html">here</a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="co"># load airway RNA sequencing data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>

<span class="co"># convert to tidybulk tibble</span>
<span class="va">counts_airway</span> <span class="op">&lt;-</span>
    <span class="va">airway</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p>We can type the name of the object to view.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="va">counts_airway</span>
<span class="co">#&gt; # A tibble: 512,816 x 12</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;fct&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… SRR10…    467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… SRR10…    260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… SRR10…     60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… SRR10…   3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… SRR10…   1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… SRR10…    519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… SRR10…    394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 512,806 more rows, and 2 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;</span></pre></div>
<p>The <code>counts_tt</code> object contains information about genes and samples, the first column has the Ensembl gene identifier, the second column has the sample identifier and the third column has the gene transcription abundance expressed in number of reads aligning to the gene in each experimental sample. The remaining columns include sample-wise information. The dex column tells us whether the samples are treated or untreated and the cell column tells us what cell line they are from.</p>
<p>We can shorten the sample names. We can remove the SRR1039 prefix that’s present in all of them, as shorter names can fit better in some of the plots we will create. We can use <code><a href="https://rdrr.io/pkg/tidyseurat/man/mutate.html">mutate()</a></code> together with <code><a href="https://stringr.tidyverse.org/reference/str_replace.html">str_replace()</a></code> to remove the SRR1039 string from the sample column.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="va">counts_format</span> <span class="op">&lt;-</span> <span class="va">counts_airway</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>sample<span class="op">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"SRR1039"</span><span class="op">)</span><span class="op">)</span></pre></div>
</div>
<div id="adding-gene-symbols" class="section level3">
<h3 class="hasAnchor">
<a href="#adding-gene-symbols" class="anchor"></a>Adding gene symbols</h3>
<p>We can get the gene symbols for these Ensembl gene ids using the Bioconductor annotation package for human, <code>org.Hs.eg.db</code> and add them as a column using <code>mutate</code> again.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="va">counts_tt</span> <span class="op">&lt;-</span> <span class="va">counts_format</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span>, keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span>, column<span class="op">=</span><span class="st">"SYMBOL"</span>, multiVals <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; 'select()' returned 1:many mapping between keys and columns</span>
<span class="va">counts_tt</span>
<span class="co">#&gt; # A tibble: 512,816 x 13</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508         0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508         0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 512,806 more rows, and 3 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;</span></pre></div>
<p>With tidyverse, all above operations can be linked with the <code><a href="https://rdrr.io/pkg/tidyseurat/man/pipe.html">%&gt;%</a></code>, as shown below. This has the benefits that</p>
<ul>
<li>no temporary variables need to be created</li>
<li>less typing is required</li>
<li>the steps can be seen more clearly.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">counts_tt</span> <span class="op">&lt;-</span>    
    <span class="va">airway</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>sample<span class="op">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"SRR1039"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">mapIds</a></span><span class="op">(</span><span class="fu">org.Hs.eg.db</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/org.Hs.eg.db/man/org.Hs.egBASE.html">org.Hs.eg.db</a></span>, keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span>, keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span>, column<span class="op">=</span><span class="st">"SYMBOL"</span>, multiVals <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; 'select()' returned 1:many mapping between keys and columns</span></pre></div>
<p>From this tidybulk tibble, we can perform differential expression analysis with the tidybulk package.</p>
</div>
</div>
<div id="filtering-lowly-transcribed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-lowly-transcribed-genes" class="anchor"></a>Filtering lowly transcribed genes</h2>
<p>Genes with very low counts across all libraries provide little evidence for differential expression and they interfere with some of the statistical approximations that are used later in the pipeline. They also add to the multiple testing burden when estimating false discovery rates, reducing power to detect differentially expressed genes. These genes should be filtered out prior to further analysis.</p>
<p>Tidybulk <code>keep_abundant</code> can use the <em>edgeR</em> <code>filterByExpr</code> function described in <span class="citation">(Law et al. 2016)</span> to keep the genes with adequate abundance for differential expression testing. By default, this will keep genes with ~10 counts in a minimum number of samples, the number of the samples in the smallest group. In this dataset the smallest group size is four (as we have four dex-treated samples vs four untreated). Alternatively, we could use <code>identify_abundant</code> to identify which genes are abundant or not (TRUE/FALSE), rather than just keeping the abundant ones.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="co"># Filtering counts</span>
<span class="va">counts_filtered</span> <span class="op">&lt;-</span> <span class="va">counts_tt</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span><span class="op">(</span>factor_of_interest<span class="op">=</span><span class="va">dex</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">counts_filtered</span>
<span class="co">#&gt; # A tibble: 127,408 x 14</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508      2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 127,398 more rows, and 4 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;</span></pre></div>
<p>After running <code>keep_abundant</code> we have a column called <code>.abundant</code> containing TRUE (<code>identify_abundant</code> would have TRUE/FALSE).</p>
</div>
<div id="scaling-counts-to-normalise" class="section level2">
<h2 class="hasAnchor">
<a href="#scaling-counts-to-normalise" class="anchor"></a>Scaling counts to normalise</h2>
<p>Scaling of counts, normalisation, is performed to eliminate uninteresting differences between samples due to sequencing depth or composition. In the tidybulk package the function <code>scale_abundance</code> generates scaled counts, with scaling factors calculated on abundant (filtered) transcripts and applied to all transcripts. We can choose from different normalisation methods. Here we will use the default, TMM, <span class="citation">(Robinson and Oshlack 2010)</span>. TMM normalisation (and most scaling normalisation methods) scale relative to one sample.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="co"># Scaling counts</span>
<span class="va">counts_scaled</span> <span class="op">&lt;-</span> <span class="va">counts_filtered</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">counts_scaled</span>
<span class="co">#&gt; # A tibble: 127,408 x 17</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508      2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 127,398 more rows, and 7 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, TMM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   multiplier &lt;dbl&gt;, counts_scaled &lt;dbl&gt;</span></pre></div>
<p>After we run <code>scale_abundance</code> we should see some columns have been added at the end. The <code>counts_scaled</code> column contains the scaled counts.</p>
<p>We can visualise the difference of abundance densities before and after scaling. As tidybulk output is compatible with tidyverse, we can simply pipe it into standard tidyverse functions such as <code>filter</code>, <code>pivot_longer</code> and <code>ggplot</code>. We can also take advantage of ggplot’s <code>facet_wrap</code> to easily create multiple plots.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"counts"</span>, <span class="st">"counts_scaled"</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"source"</span>, values_to <span class="op">=</span> <span class="st">"abundance"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">abundance</span> <span class="op">+</span> <span class="fl">1</span>, color<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">source</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="va">custom_theme</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-12-1.png" width="70%"></p>
<p>In this dataset the distributions of the counts are not very different to each other before scaling but scaling does make the distributions more similar. If we saw a sample with a very different distribution we may need to investigate it.</p>
<p>As tidybulk smoothly integrates with ggplot2 and other tidyverse packages it can save on typing and make plots easier to generate. Compare the code for creating density plots with tidybulk versus standard base R below (standard code adapted from <span class="citation">(Law et al. 2016)</span>).</p>
<p><strong>tidybulk</strong></p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="co"># tidybulk</span>
<span class="va">airway</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span><span class="op">(</span>factor_of_interest<span class="op">=</span><span class="va">dex</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"counts"</span>, <span class="st">"counts_scaled"</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"source"</span>, values_to <span class="op">=</span> <span class="st">"abundance"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">abundance</span> <span class="op">+</span> <span class="fl">1</span>, color<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">source</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="va">custom_theme</span></pre></div>
<p><strong>base R using edgeR</strong></p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="co"># Example code, no need to run</span>

<span class="co"># Prepare data set</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu">SE2DGEList</span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu">filterByExpr</span><span class="op">(</span><span class="va">dgList</span>, group<span class="op">=</span><span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>,, keep.lib.sizes<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span>
<span class="va">nsamples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>

<span class="co"># Setup graphics</span>
<span class="va">col</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="va">nsamples</span>, <span class="st">"Paired"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Plot raw counts</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span>, col<span class="op">=</span><span class="va">col</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, lwd<span class="op">=</span><span class="fl">2</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.26</span><span class="op">)</span>, las<span class="op">=</span><span class="fl">2</span>, main<span class="op">=</span><span class="st">""</span>, xlab<span class="op">=</span><span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span>main<span class="op">=</span><span class="st">"Counts"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="va">nsamples</span><span class="op">)</span><span class="op">{</span>
  <span class="va">den</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">den</span><span class="op">$</span><span class="va">x</span>, <span class="va">den</span><span class="op">$</span><span class="va">y</span>, col<span class="op">=</span><span class="va">col</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend<span class="op">=</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Run</span>, text.col<span class="op">=</span><span class="va">col</span>, bty<span class="op">=</span><span class="st">"n"</span><span class="op">)</span>

<span class="co"># Plot scaled counts</span>
<span class="va">dgList_norm</span> <span class="op">&lt;-</span> <span class="fu">calcNormFactors</span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">lcpm_n</span> <span class="op">&lt;-</span> <span class="fu">cpm</span><span class="op">(</span><span class="va">dgList_norm</span>, log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">lcpm_n</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span>, col<span class="op">=</span><span class="va">col</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, lwd<span class="op">=</span><span class="fl">2</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.26</span><span class="op">)</span>, las<span class="op">=</span><span class="fl">2</span>, main<span class="op">=</span><span class="st">""</span>, xlab<span class="op">=</span><span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="st">"Counts scaled"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="va">nsamples</span><span class="op">)</span><span class="op">{</span>
  <span class="va">den</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">lcpm_n</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">den</span><span class="op">$</span><span class="va">x</span>, <span class="va">den</span><span class="op">$</span><span class="va">y</span>, col<span class="op">=</span><span class="va">col</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend<span class="op">=</span><span class="va">dgList_norm</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Run</span>, text.col<span class="op">=</span><span class="va">col</span>, bty<span class="op">=</span><span class="st">"n"</span><span class="op">)</span></pre></div>
</div>
<div id="exploratory-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#exploratory-analyses" class="anchor"></a>Exploratory analyses</h2>
<div id="dimensionality-reduction" class="section level3">
<h3 class="hasAnchor">
<a href="#dimensionality-reduction" class="anchor"></a>Dimensionality reduction</h3>
<p>By far, one of the most important plots we make when we analyse RNA sequencing data are principal-component analysis (PCA) or multi-dimensional scaling (MDS) plots. We reduce the dimensions of the data to identify the greatest sources of variation in the data. A principal components analysis is an example of an unsupervised analysis, where we don’t need to specify the groups. If your experiment is well controlled and has worked well, what we hope to see is that the greatest sources of variation in the data are the treatments/groups we are interested in. It is also an incredibly useful tool for quality control and checking for outliers. We can use the <code>reduce_dimensions</code> function to calculate the dimensions.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="co"># Get principal components</span>
<span class="va">counts_scal_PCA</span> <span class="op">&lt;-</span>
  <span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span>
<span class="co">#&gt; Getting the 500 most variable genes</span>
<span class="co">#&gt; Fraction of variance explained by the selected principal components</span>
<span class="co">#&gt; # A tibble: 2 x 2</span>
<span class="co">#&gt;   `Fraction of variance`    PC</span>
<span class="co">#&gt;                    &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1                  0.355     1</span>
<span class="co">#&gt; 2                  0.309     2</span>
<span class="co">#&gt; tidybulk says: to access the raw results do `attr(..., "internals")$PCA`</span></pre></div>
<p>This joins the result to the counts object.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="co"># Take a look</span>
<span class="va">counts_scal_PCA</span>
<span class="co">#&gt; # A tibble: 127,408 x 19</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508      2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 127,398 more rows, and 9 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, TMM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   multiplier &lt;dbl&gt;, counts_scaled &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;</span></pre></div>
<p>For plotting, we can select just the sample-wise information with <code>pivot_sample</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="co"># take a look</span>
<span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 8 x 15</span>
<span class="co">#&gt;   sample SampleName cell  dex   albut Run   avgLength Experiment Sample</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; </span>
<span class="co">#&gt; 1 508    GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; 2 509    GSM1275863 N613… trt   untrt SRR1…       126 SRX384346  SRS50…</span>
<span class="co">#&gt; 3 512    GSM1275866 N052… untrt untrt SRR1…       126 SRX384349  SRS50…</span>
<span class="co">#&gt; 4 513    GSM1275867 N052… trt   untrt SRR1…        87 SRX384350  SRS50…</span>
<span class="co">#&gt; 5 516    GSM1275870 N080… untrt untrt SRR1…       120 SRX384353  SRS50…</span>
<span class="co">#&gt; 6 517    GSM1275871 N080… trt   untrt SRR1…       126 SRX384354  SRS50…</span>
<span class="co">#&gt; 7 520    GSM1275874 N061… untrt untrt SRR1…       101 SRX384357  SRS50…</span>
<span class="co">#&gt; 8 521    GSM1275875 N061… trt   untrt SRR1…        98 SRX384358  SRS50…</span>
<span class="co">#&gt; # … with 6 more variables: BioSample &lt;fct&gt;, .abundant &lt;lgl&gt;, TMM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   multiplier &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;</span></pre></div>
<p>We can now plot the reduced dimensions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="co"># PCA plot</span>
<span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC1</span>, y<span class="op">=</span><span class="va">PC2</span>, colour<span class="op">=</span><span class="va">dex</span>, shape<span class="op">=</span><span class="va">cell</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">sample</span><span class="op">)</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
    <span class="va">custom_theme</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-18-1.png" width="70%"></p>
<p>The samples separate by treatment on PC1 which is what we hope to see. PC2 separates the N080611 cell line from the other samples, indicating a greater difference between that cell line and the others.</p>
</div>
<div id="hierarchical-clustering-with-heatmaps" class="section level3">
<h3 class="hasAnchor">
<a href="#hierarchical-clustering-with-heatmaps" class="anchor"></a>Hierarchical clustering with heatmaps</h3>
<p>An alternative to principal component analysis for examining relationships between samples is using hierarchical clustering. Heatmaps are a nice visualisation to examine hierarchical clustering of your samples. tidybulk has a simple function we can use, <code>keep_variable</code>, to extract the most variable genes which we can then plot with tidyHeatmap.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="va">counts_scaled</span> <span class="op">%&gt;%</span>

    <span class="co"># extract 500 most variable genes</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_variable-methods.html">keep_variable</a></span><span class="op">(</span> .abundance <span class="op">=</span> <span class="va">counts_scaled</span>, top <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">%&gt;%</span>

    <span class="co"># create heatmap</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/heatmap-methods.html">heatmap</a></span><span class="op">(</span>
          .column <span class="op">=</span> <span class="va">sample</span>,
          .row <span class="op">=</span> <span class="va">feature</span>,
          .value <span class="op">=</span> <span class="va">counts_scaled</span>,
          transform <span class="op">=</span> <span class="va">log1p</span>
    <span class="op">)</span> <span class="op">%&gt;%</span>
        <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/add_tile-methods.html">add_tile</a></span><span class="op">(</span><span class="va">dex</span><span class="op">)</span> <span class="op">%&gt;%</span>
        <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/add_tile-methods.html">add_tile</a></span><span class="op">(</span><span class="va">cell</span><span class="op">)</span>
<span class="co">#&gt; Getting the 500 most variable genes</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-19-1.png" width="70%"></p>
<p>In the heatmap we can see the samples cluster into two groups, treated and untreated, for three of the cell lines, and the cell line (N080611) again is further away from the others.</p>
<p>Tidybulk enables a simplified way of generating a clustered heatmap of variable genes. Compare the code below for tidybulk versus a base R method.</p>
<p><strong>base R using edgeR</strong></p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="co"># Example code, no need to run</span>

<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu">SE2DGEList</span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu">filterByExpr</span><span class="op">(</span><span class="va">dgList</span>, group<span class="op">=</span><span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>,, keep.lib.sizes<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu">calcNormFactors</span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu">cpm</span><span class="op">(</span><span class="va">dgList</span>, log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="fl">1</span>, <span class="va">var</span><span class="op">)</span>
<span class="va">select_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">var_genes</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span>
<span class="va">highly_variable_lcpm</span> <span class="op">&lt;-</span> <span class="va">logcounts</span><span class="op">[</span><span class="va">select_var</span>,<span class="op">]</span>
<span class="va">colours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#fefada"</span> <span class="op">)</span>
<span class="va">col.group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>,<span class="st">"grey"</span><span class="op">)</span><span class="op">[</span><span class="va">group</span><span class="op">]</span>
<span class="fu">gplots</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html">heatmap.2</a></span><span class="op">(</span><span class="va">highly_variable_lcpm</span>, col<span class="op">=</span><span class="va">colours</span>, trace<span class="op">=</span><span class="st">"none"</span>, ColSideColors<span class="op">=</span><span class="va">col.group</span>, scale<span class="op">=</span><span class="st">"row"</span><span class="op">)</span></pre></div>
</div>
</div>
<div id="differential-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression" class="anchor"></a>Differential expression</h2>
<p>Now that we are happy that the data looks good, we can continue to testing for differentially expressed (DE) genes. We will use the <code>test_differential_abundance</code> from tidybulk which currently uses <em>edgeR</em> <span class="citation">(Robinson, McCarthy, and Smyth 2010)</span> to perform the differential expression analysis. We give <code>test_differential_abundance</code> our tidybulk counts object and a formula, specifying the column that contains our groups to be compared. If all our samples were from the same cell line we could use the formula <code>0 + dex</code>, however, each treated and untreated sample is from a different cell line so we add the cell line as an additional factor <code>0 + dex + cell</code>. We also provide the names of the groups we want to compare to <code>.contrasts</code> (e.g. <code>.contrasts = c("dextreat - dexuntreat")</code>).</p>
<p>We only have one contrast here so we omit the suffix.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="va">counts_de</span> <span class="op">&lt;-</span> <span class="va">counts_filtered</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
      .formula <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
      .contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"dextrt - dexuntrt"</span><span class="op">)</span>,
      omit_contrast_in_colnames <span class="op">=</span> <span class="cn">TRUE</span>
    <span class="op">)</span>
<span class="co">#&gt; tidybulk says: All methods use raw counts, </span>
<span class="co">#&gt; irrespective of if scale_abundance or adjust_abundance have been calculated, </span>
<span class="co">#&gt; therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</span>
<span class="co">#&gt; tidybulk says: The design column names are "dextrt, dexuntrt, cellN061011, cellN080611, cellN61311"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span></pre></div>
<p>The results will be joined to our counts for every sample.</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="co"># take a look</span>
<span class="va">counts_de</span>
<span class="co">#&gt; # A tibble: 127,408 x 19</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508      2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 127,398 more rows, and 9 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, logFC &lt;dbl&gt;, logCPM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   F &lt;dbl&gt;, PValue &lt;dbl&gt;, FDR &lt;dbl&gt;</span></pre></div>
<p>If we just want a table of differentially expressed genes we can select the transcript-wise information with <code>pivot_transcript</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="co"># take a look</span>
<span class="va">counts_de</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 15,926 x 9</span>
<span class="co">#&gt;    feature       albut symbol   .abundant   logFC logCPM       F  PValue     FDR</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;fct&gt; &lt;chr&gt;    &lt;lgl&gt;       &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1 ENSG00000000… untrt TSPAN6   TRUE      -0.390    5.06 32.8    3.12e-4 0.00283</span>
<span class="co">#&gt;  2 ENSG00000000… untrt DPM1     TRUE       0.198    4.61  6.90   2.81e-2 0.0770 </span>
<span class="co">#&gt;  3 ENSG00000000… untrt SCYL3    TRUE       0.0292   3.48  0.0969 7.63e-1 0.844  </span>
<span class="co">#&gt;  4 ENSG00000000… untrt C1orf112 TRUE      -0.124    1.47  0.377  5.55e-1 0.682  </span>
<span class="co">#&gt;  5 ENSG00000000… untrt CFH      TRUE       0.417    8.09 29.3    4.63e-4 0.00376</span>
<span class="co">#&gt;  6 ENSG00000001… untrt FUCA2    TRUE      -0.250    5.91 14.9    4.05e-3 0.0186 </span>
<span class="co">#&gt;  7 ENSG00000001… untrt GCLC     TRUE      -0.0581   4.84  0.167  6.92e-1 0.794  </span>
<span class="co">#&gt;  8 ENSG00000001… untrt NFYA     TRUE      -0.509    4.13 44.9    1.00e-4 0.00126</span>
<span class="co">#&gt;  9 ENSG00000001… untrt STPG1    TRUE      -0.136    3.12  1.04   3.35e-1 0.478  </span>
<span class="co">#&gt; 10 ENSG00000001… untrt NIPAL3   TRUE      -0.0500   7.04  0.350  5.69e-1 0.695  </span>
<span class="co">#&gt; # … with 15,916 more rows</span></pre></div>
<p>Now we have columns with our log-fold change (logFC), false-discovery rate (FDR) and probability value (p-value).</p>
<p>Tidybulk enables a simplified way of performing an RNA sequencing differential expression analysis (with the benefit of smoothly integrating with ggplot2 and other tidyverse packages). Compare the code for a tidybulk edgeR analysis versus standard edgeR below.</p>
<p><strong>standard edgeR</strong></p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="co"># Example code, no need to run</span>

<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu">SE2DGEList</span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu">filterByExpr</span><span class="op">(</span><span class="va">dgList</span>, group<span class="op">=</span><span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>,, keep.lib.sizes<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu">calcNormFactors</span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span>
<span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">group</span> <span class="op">+</span> <span class="va">cell</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu">estimateDisp</span><span class="op">(</span><span class="va">dgList</span>, <span class="va">design</span><span class="op">)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">glmQLFit</span><span class="op">(</span><span class="va">dgList</span>, <span class="va">design</span><span class="op">)</span>
<span class="va">TvsU</span> <span class="op">&lt;-</span> <span class="fu">makeContrasts</span><span class="op">(</span>TvsU<span class="op">=</span><span class="va">grouptrt</span><span class="op">-</span><span class="va">groupuntrt</span>, levels<span class="op">=</span><span class="va">design</span><span class="op">)</span>
<span class="va">qlf</span> <span class="op">&lt;-</span> <span class="fu">glmQLFTest</span><span class="op">(</span><span class="va">fit</span>, contrast<span class="op">=</span><span class="va">TvsU</span><span class="op">)</span></pre></div>
<div id="table-of-differentially-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#table-of-differentially-expressed-genes" class="anchor"></a>Table of differentially expressed genes</h3>
<p>We can write out our differentially expressed genes to a file that can be loaded into e.g. Excel. <code>write_tsv</code> will create a tab-separated file.</p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="co"># save results</span>
<span class="va">counts_de</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv</a></span><span class="op">(</span><span class="st">"de_results.tsv"</span><span class="op">)</span></pre></div>
</div>
<div id="counting-differentially-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#counting-differentially-expressed-genes" class="anchor"></a>Counting differentially expressed genes</h3>
<p>We can count how many differentially expressed genes there are. We’ll filter on FDR 0.05.</p>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="va">counts_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>num_de <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html">n_distinct</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1 x 1</span>
<span class="co">#&gt;   num_de</span>
<span class="co">#&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   4967</span></pre></div>
</div>
<div id="extracting-top-differentially-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#extracting-top-differentially-expressed-genes" class="anchor"></a>Extracting top differentially expressed genes</h3>
<p>We can see the top genes by smallest p-value. We’ll take a look at the top 6.</p>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="va">topgenes</span> <span class="op">&lt;-</span>
    <span class="va">counts_de</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">PValue</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span>

<span class="va">topgenes</span>
<span class="co">#&gt; # A tibble: 6 x 9</span>
<span class="co">#&gt;   feature         albut symbol  .abundant logFC logCPM     F   PValue        FDR</span>
<span class="co">#&gt;   &lt;chr&gt;           &lt;fct&gt; &lt;chr&gt;   &lt;lgl&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1 ENSG00000165995 untrt CACNB2  TRUE       3.28   4.51 1575. 3.34e-11    4.07e-7</span>
<span class="co">#&gt; 2 ENSG00000109906 untrt ZBTB16  TRUE       7.15   4.15 1429. 5.11e-11    4.07e-7</span>
<span class="co">#&gt; 3 ENSG00000146250 untrt PRSS35  TRUE      -2.76   3.91  807. 6.16e-10    2.57e-6</span>
<span class="co">#&gt; 4 ENSG00000162493 untrt PDPN    TRUE       1.88   5.68  768. 7.60e-10    2.57e-6</span>
<span class="co">#&gt; 5 ENSG00000152583 untrt SPARCL1 TRUE       4.56   5.53  721. 1.00e- 9    2.57e-6</span>
<span class="co">#&gt; 6 ENSG00000120129 untrt DUSP1   TRUE       2.94   7.31  694. 1.18e- 9    2.57e-6</span></pre></div>
<p>We can extract the symbols for these top genes to use in some of the plots we will make.</p>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="va">topgenes_symbols</span> <span class="op">&lt;-</span> <span class="va">topgenes</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">topgenes_symbols</span>
<span class="co">#&gt; ENSG00000165995 ENSG00000109906 ENSG00000146250 ENSG00000162493 ENSG00000152583 </span>
<span class="co">#&gt;        "CACNB2"        "ZBTB16"        "PRSS35"          "PDPN"       "SPARCL1" </span>
<span class="co">#&gt; ENSG00000120129 </span>
<span class="co">#&gt;         "DUSP1"</span></pre></div>
</div>
</div>
<div id="plots-after-testing-for-differentially-expressed" class="section level2">
<h2 class="hasAnchor">
<a href="#plots-after-testing-for-differentially-expressed" class="anchor"></a>Plots after testing for differentially expressed</h2>
<div id="volcano-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#volcano-plots" class="anchor"></a>Volcano plots</h3>
<p>Volcano plots are a useful genome-wide plot for checking that the analysis looks good. Volcano plots enable us to visualise the significance of change (p-value) versus the fold change (logFC). Highly significant genes are towards the top of the plot. We can also colour significant genes (e.g. genes with false-discovery rate &lt; 0.05)</p>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="co"># volcano plot, minimal</span>
<span class="va">counts_de</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">logFC</span>, y<span class="op">=</span><span class="va">PValue</span>, colour<span class="op">=</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"log10_reverse"</span><span class="op">)</span> <span class="op">+</span>
    <span class="va">custom_theme</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-29-1.png" width="70%"></p>
<p>A more informative plot, integrating some of the packages in tidyverse.</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="va">counts_de</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Subset data</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>significant <span class="op">=</span> <span class="va">FDR</span><span class="op">&lt;</span><span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;=</span><span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">symbol</span> <span class="op">%in%</span> <span class="va">topgenes_symbols</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>, <span class="st">""</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Plot</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span>, label<span class="op">=</span><span class="va">symbol</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">significant</span>, size <span class="op">=</span> <span class="va">significant</span>, alpha<span class="op">=</span><span class="va">significant</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>

    <span class="co"># Custom scales</span>
    <span class="va">custom_theme</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"log10_reverse"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"#e11f28"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Scale for 'colour' is already present. Adding another scale for 'colour',</span>
<span class="co">#&gt; which will replace the existing scale.</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-30-1.png" width="70%"></p>
</div>
<div id="stripcharts" class="section level3">
<h3 class="hasAnchor">
<a href="#stripcharts" class="anchor"></a>Stripcharts</h3>
<p>Before following up on the differentially expressed genes with further lab work, it is also recommended to have a look at the expression levels of the individual samples for the genes of interest. We can use stripcharts to do this. These will help show if expression is consistent amongst replicates in the groups.</p>
<p>With stripcharts we can see if replicates tend to group together and how the expression compares to the other groups. We’ll also add a box plot to show the distribution.</p>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="va">strip_chart</span> <span class="op">&lt;-</span>
    <span class="va">counts_scaled</span> <span class="op">%&gt;%</span>

    <span class="co"># extract counts for top differentially expressed genes</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">symbol</span> <span class="op">%in%</span> <span class="va">topgenes_symbols</span><span class="op">)</span> <span class="op">%&gt;%</span>

    <span class="co"># make faceted stripchart</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dex</span>, y <span class="op">=</span> <span class="va">counts_scaled</span> <span class="op">+</span> <span class="fl">1</span>, fill <span class="op">=</span> <span class="va">dex</span>, label <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">symbol</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span>
    <span class="va">custom_theme</span>

<span class="va">strip_chart</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-31-1.png" width="70%"></p>
<p>We can also easily check the raw and scaled counts for these genes.</p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="va">counts_scaled</span> <span class="op">%&gt;%</span>

  <span class="co"># extract counts for top differentially expressed genes</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">symbol</span> <span class="op">%in%</span> <span class="va">topgenes_symbols</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># reshape to create column ("source") containing the raw and scaled counts     </span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">counts</span>, <span class="va">counts_scaled</span><span class="op">)</span>, 
    names_to <span class="op">=</span> <span class="st">"source"</span>, 
    values_to <span class="op">=</span> <span class="st">"count"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># make faceted stripchart</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">source</span>, y <span class="op">=</span> <span class="va">count</span> <span class="op">+</span> <span class="fl">1</span>, fill <span class="op">=</span> <span class="va">dex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">symbol</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="va">custom_theme</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
</div>
</div>
<div id="interactive-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#interactive-plots" class="anchor"></a>Interactive Plots</h2>
<p>A really nice feature of using tidyverse and ggplot2 is that we can make interactive plots quite easily using the plotly package. This can be very useful for exploring what genes or samples are in the plots. We can make interactive plots directly from our ggplot2 object (strip_chart). Having <code>label</code> in the <code>aes</code> is useful to visualise the identifier of the data point (here the sample id) or other variables when we hover over the plot.</p>
<p>We can also specify which parameters from the <code>aes</code> we want to show up when we hover over the plot with <code>tooltip</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="va">strip_chart</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/plotly/man/ggplotly.html">ggplotly</a></span><span class="op">(</span>tooltip <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"label"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">)</span></pre></div>
<div id="htmlwidget-1ecd539de4e3c212fe3c" style="width:70%;height:432.632880098888px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-1ecd539de4e3c212fe3c">{"x":{"data":[{"x":[1,1,1,1],"y":[2.69128969599625,3.07349503218585,2.89433133924867,3.00332359018516],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,1)","legendgroup":"(trt,1)","showlegend":true,"xaxis":"x","yaxis":"y","frame":null},{"x":[1,1,1,1],"y":[3.73064415538561,3.81600288632387,3.63773959490878,3.84413376382407],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,2)","legendgroup":"(trt,2)","showlegend":true,"xaxis":"x2","yaxis":"y","frame":null},{"x":[1,1,1,1],"y":[3.20827199577471,3.48606374326543,2.82530745673601,3.12027108528531],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,3)","legendgroup":"(trt,3)","showlegend":true,"xaxis":"x3","yaxis":"y","frame":null},{"x":[1,1,1,1],"y":[1.65242137242843,2.11154731277845,1.93112316960655,1.79419901680174],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,4)","legendgroup":"(trt,4)","showlegend":true,"xaxis":"x","yaxis":"y2","frame":null},{"x":[1,1,1,1],"y":[3.33968854733852,3.227969656062,3.12063282561804,3.33922209086478],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,5)","legendgroup":"(trt,5)","showlegend":true,"xaxis":"x2","yaxis":"y2","frame":null},{"x":[1,1,1,1],"y":[2.89905240413481,2.79194409457512,2.60136364157697,3.05589265897356],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,6)","legendgroup":"(trt,6)","showlegend":true,"xaxis":"x3","yaxis":"y2","frame":null},{"x":[2,2,2,2],"y":[1.88450930407812,2.08667246288489,1.75347199448472,2.00731696964479],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,1)","legendgroup":"(untrt,1)","showlegend":true,"xaxis":"x","yaxis":"y","frame":null},{"x":[2,2,2,2],"y":[2.79922778113043,3.00007848466273,2.8633343353318,2.82557281829644],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,2)","legendgroup":"(untrt,2)","showlegend":true,"xaxis":"x2","yaxis":"y","frame":null},{"x":[2,2,2,2],"y":[2.92610890566171,2.62268358131662,2.25013531982859,2.57579220801279],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,3)","legendgroup":"(untrt,3)","showlegend":true,"xaxis":"x3","yaxis":"y","frame":null},{"x":[2,2,2,2],"y":[2.49561949505115,2.87397144019785,2.77130167407313,2.6641608293693],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,4)","legendgroup":"(untrt,4)","showlegend":true,"xaxis":"x","yaxis":"y2","frame":null},{"x":[2,2,2,2],"y":[1.77545286096872,1.91671162426974,1.91850446396388,1.9235908013111],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,5)","legendgroup":"(untrt,5)","showlegend":true,"xaxis":"x2","yaxis":"y2","frame":null},{"x":[2,2,2,2],"y":[0.679652690730562,1.13313583033034,0.259809882187156,0.709154121013273],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,6)","legendgroup":"(untrt,6)","showlegend":true,"xaxis":"x3","yaxis":"y2","frame":null},{"x":[0.810440228879452,1.01889476515353,0.866175622306764,0.727235584333539],"y":[2.89431102733781,3.07352808877074,2.69127719115771,3.00339277197431],"text":["counts_scaled + 1:  784.027578<br />sample: 509","counts_scaled + 1: 1184.390819<br />sample: 513","counts_scaled + 1:  491.235445<br />sample: 517","counts_scaled + 1: 1007.682207<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,1)","legendgroup":"(trt,1)","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.33582283519208,1.33958725780249,1.39736013840884,0.622571564465761],"y":[3.730565861186,3.8160507495562,3.63773602042425,3.84421307550546],"text":["counts_scaled + 1: 5378.289251<br />sample: 509","counts_scaled + 1: 6546.405248<br />sample: 513","counts_scaled + 1: 4342.497678<br />sample: 517","counts_scaled + 1: 6984.474946<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,2)","legendgroup":"(trt,2)","showlegend":false,"xaxis":"x2","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.38495596479625,1.04086741600186,0.903692658618093,1.16279125325382],"y":[3.48613227901004,3.20834830290181,2.82537256656381,3.1202127641323],"text":["counts_scaled + 1: 3062.412884<br />sample: 509","counts_scaled + 1: 1615.369936<br />sample: 513","counts_scaled + 1:  668.817236<br />sample: 517","counts_scaled + 1: 1319.079847<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,3)","legendgroup":"(trt,3)","showlegend":false,"xaxis":"x3","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.900436445325613,1.30361659433693,0.965372990444303,1.0263085545972],"y":[1.65246746343072,2.11149569633156,1.9312236661858,1.79422529775535],"text":["counts_scaled + 1:   44.918099<br />sample: 509","counts_scaled + 1:  129.284754<br />sample: 513","counts_scaled + 1:   85.334209<br />sample: 517","counts_scaled + 1:   62.258552<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,4)","legendgroup":"(trt,4)","showlegend":false,"xaxis":"x","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[1.30412703529,0.973745271191001,1.04116196446121,0.948173537664115],"y":[3.33978768719375,3.22793553687133,3.120535447651,3.33931057282865],"text":["counts_scaled + 1: 2186.193241<br />sample: 509","counts_scaled + 1: 1690.322826<br />sample: 513","counts_scaled + 1: 1320.179017<br />sample: 517","counts_scaled + 1: 2183.846409<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,5)","legendgroup":"(trt,5)","showlegend":false,"xaxis":"x2","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[0.907581271789968,1.29034097529948,1.16939483545721,0.671565954759717],"y":[2.8991357809772,2.79186630914142,2.6014190057022,3.05587148329046],"text":["counts_scaled + 1:  792.596963<br />sample: 509","counts_scaled + 1:  619.361341<br />sample: 513","counts_scaled + 1:  399.359152<br />sample: 517","counts_scaled + 1: 1137.346143<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,6)","legendgroup":"(trt,6)","showlegend":false,"xaxis":"x3","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[2.04948859475553,2.28580316919833,1.71540137603879,2.09838115870953],"y":[1.88455994266574,2.08662556432192,1.75342109747052,2.00741634847938],"text":["counts_scaled + 1:   76.649496<br />sample: 508","counts_scaled + 1:  122.087855<br />sample: 512","counts_scaled + 1:   56.685502<br />sample: 516","counts_scaled + 1:  101.699067<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,1)","legendgroup":"(untrt,1)","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2.24471744187176,1.80914960876107,2.28455028701574,2.20718590579927],"y":[2.7993225131933,3.00001065509452,2.82558553221706,2.86337105147168],"text":["counts_scaled + 1:  629.836436<br />sample: 508","counts_scaled + 1: 1000.180734<br />sample: 512","counts_scaled + 1:  669.226020<br />sample: 516","counts_scaled + 1:  730.019289<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,2)","legendgroup":"(untrt,2)","showlegend":false,"xaxis":"x2","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.67830618210137,2.16890415400267,2.19833790063858,1.83517817631364],"y":[2.9261480159247,2.6227136888317,2.25015293282736,2.57578654714036],"text":["counts_scaled + 1:  843.546263<br />sample: 508","counts_scaled + 1:  419.453267<br />sample: 512","counts_scaled + 1:  177.883358<br />sample: 516","counts_scaled + 1:  376.523605<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,3)","legendgroup":"(untrt,3)","showlegend":false,"xaxis":"x3","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.70950626730919,1.99382490273565,1.86888039149344,2.12575019150972],"y":[2.49563976596165,2.87386733927495,2.77133552108418,2.66426198523135],"text":["counts_scaled + 1:  313.054171<br />sample: 508","counts_scaled + 1:  748.120301<br />sample: 512","counts_scaled + 1:  590.611194<br />sample: 516","counts_scaled + 1:  461.488443<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,4)","legendgroup":"(untrt,4)","showlegend":false,"xaxis":"x","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[2.36305609457195,2.12314799614251,1.80739749539644,1.62399844154716],"y":[1.77539513988024,1.91670793316536,1.91840716336147,1.92358144805956],"text":["counts_scaled + 1:   59.628359<br />sample: 508","counts_scaled + 1:   82.548963<br />sample: 512","counts_scaled + 1:   82.890444<br />sample: 516","counts_scaled + 1:   83.866941<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,5)","legendgroup":"(untrt,5)","showlegend":false,"xaxis":"x2","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[1.95647639483213,2.23195655494928,2.00140024814755,1.69934156164527],"y":[0.679600394842287,0.709127288989701,0.259712724038636,1.13308673315812],"text":["counts_scaled + 1:    4.782475<br />sample: 508","counts_scaled + 1:    5.118635<br />sample: 512","counts_scaled + 1:    1.818904<br />sample: 516","counts_scaled + 1:   13.587383<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,6)","legendgroup":"(untrt,6)","showlegend":false,"xaxis":"x3","yaxis":"y2","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":40.5768157463462,"r":7.30593607305936,"b":44.1700079422781,"l":58.7795765877958},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":15.9402241594022},"xaxis":{"domain":[0,0.32941943900848],"automargin":true,"type":"linear","autorange":false,"range":[0.4,2.6],"tickmode":"array","ticktext":["trt","untrt"],"tickvals":[1,2],"categoryorder":"array","categoryarray":["trt","untrt"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":12.7521793275218},"tickangle":-30,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.265670402656704,"zeroline":false,"anchor":"y2","title":"","hoverformat":".2f"},"annotations":[{"text":"dex","x":0.5,"y":-0.0211346180148144,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":15.9402241594022},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"top","annotationType":"axis"},{"text":"counts_scaled + 1","x":-0.0233054616616261,"y":0.5,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":15.9402241594022},"xref":"paper","yref":"paper","textangle":-90,"xanchor":"right","yanchor":"center","annotationType":"axis"},{"text":"CACNB2","x":0.16470971950424,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"DUSP1","x":0.5,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"PDPN","x":0.83529028049576,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"PRSS35","x":0.16470971950424,"y":0.482608570461924,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"SPARCL1","x":0.5,"y":0.482608570461924,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"ZBTB16","x":0.83529028049576,"y":0.482608570461924,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"dex","x":1.02,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":15.9402241594022},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"left","yanchor":"bottom","legendTitle":true}],"yaxis":{"domain":[0.517391429538076,1],"automargin":true,"type":"linear","autorange":false,"range":[0.0804877064652949,4.02343809307881],"tickmode":"array","ticktext":["10","100","1000","10000"],"tickvals":[1,2,3,4],"categoryorder":"array","categoryarray":["10","100","1000","10000"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":12.7521793275218},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.265670402656704,"zeroline":false,"anchor":"x","title":"","hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0.517391429538076,"y1":1},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0,"y1":24.4416770444168,"yanchor":1,"ysizemode":"pixel"},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0.517391429538076,"y1":1},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0,"y1":24.4416770444168,"yanchor":1,"ysizemode":"pixel"},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0.517391429538076,"y1":1},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0,"y1":24.4416770444168,"yanchor":1,"ysizemode":"pixel"},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0,"y1":0.482608570461924},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0,"y1":24.4416770444168,"yanchor":0.482608570461924,"ysizemode":"pixel"},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0,"y1":0.482608570461924},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0,"y1":24.4416770444168,"yanchor":0.482608570461924,"ysizemode":"pixel"},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0,"y1":0.482608570461924},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0,"y1":24.4416770444168,"yanchor":0.482608570461924,"ysizemode":"pixel"}],"xaxis2":{"type":"linear","autorange":false,"range":[0.4,2.6],"tickmode":"array","ticktext":["trt","untrt"],"tickvals":[1,2],"categoryorder":"array","categoryarray":["trt","untrt"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":12.7521793275218},"tickangle":-30,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":true,"domain":[0.337247227658187,0.662752772341813],"gridcolor":"rgba(235,235,235,1)","gridwidth":0.265670402656704,"zeroline":false,"anchor":"y2","title":"","hoverformat":".2f"},"xaxis3":{"type":"linear","autorange":false,"range":[0.4,2.6],"tickmode":"array","ticktext":["trt","untrt"],"tickvals":[1,2],"categoryorder":"array","categoryarray":["trt","untrt"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":12.7521793275218},"tickangle":-30,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":true,"domain":[0.67058056099152,1],"gridcolor":"rgba(235,235,235,1)","gridwidth":0.265670402656704,"zeroline":false,"anchor":"y2","title":"","hoverformat":".2f"},"yaxis2":{"type":"linear","autorange":false,"range":[0.0804877064652949,4.02343809307881],"tickmode":"array","ticktext":["10","100","1000","10000"],"tickvals":[1,2,3,4],"categoryorder":"array","categoryarray":["10","100","1000","10000"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":12.7521793275218},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":true,"domain":[0,0.482608570461924],"gridcolor":"rgba(235,235,235,1)","gridwidth":0.265670402656704,"zeroline":false,"anchor":"x","title":"","hoverformat":".2f"},"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":12.7521793275218},"y":0.960675435801739},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"206e4d50e038":{"x":{},"y":{},"fill":{},"label":{},"type":"box"},"206e7ef1be4b":{"x":{},"y":{},"fill":{},"label":{}}},"cur_data":"206e4d50e038","visdat":{"206e4d50e038":["function (y) ","x"],"206e7ef1be4b":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<div id="key-points" class="section level2">
<h2 class="hasAnchor">
<a href="#key-points" class="anchor"></a>Key Points</h2>
<ul>
<li>RNA sequencing data can be represented and analysed in a ‘tidy’ way using tidybulk and the tidyverse</li>
<li>With the modularity offered by piping we don’t need to create variables, unless an object is used more than one. This improves robustness of the code.</li>
<li>The principles of tidy transcriptomics are to interface as much as possible with commonly known manipulation and visualisation tools, rather than creating custom functions.</li>
<li>Some of the key steps in an RNA sequencing analysis are (i) filtering lowly abundant transcripts, (ii) adjusting for differences in sequencing depth and composition, (iii) testing for differential expression</li>
<li>Dimensionality reduction (PCA or MDS) plots are very important for exploring the data</li>
<li>Density plots, volcano plots, strip-charts and heatmaps are useful visualisation tools for evaluating the hypothesis testing.</li>
</ul>
</div>
<div id="supplementary" class="section level2">
<h2 class="hasAnchor">
<a href="#supplementary" class="anchor"></a>Supplementary</h2>
<p>Some things we don’t have time to cover in Part 1 of this workshop can be found in the <a href="https://stemangiola.github.io/rpharma2020_tidytranscriptomics/articles/supplementary.html">Supplementary material</a>.</p>
</div>
<div id="exercises" class="section level2">
<h2 class="hasAnchor">
<a href="#exercises" class="anchor"></a>Exercises</h2>
<p>Try to apply what you’ve learned to another dataset. This dataset was generated from the pasilla package, which obtained the data from the paper by <span class="citation">(Brooks et al. 2011)</span>. Here we provide it as a SummarizedExperiment object. The dataset has 7 samples from Drosophila (fruitfly): 3 treated with siRNA knockdown of the pasilla gene and 4 untreated controls. Load the data and create the tidybulk object with:</p>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"pasilla"</span>, package <span class="op">=</span> <span class="st">"rpharma2020tidytranscriptomics"</span><span class="op">)</span>
<span class="va">counts_tt</span> <span class="op">&lt;-</span>
    <span class="va">pasilla</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Dm.eg.db</span>, keys<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span>, keytype <span class="op">=</span> <span class="st">"FLYBASE"</span>, column<span class="op">=</span><span class="st">"SYMBOL"</span>, multiVals <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; 'select()' returned 1:1 mapping between keys and columns</span></pre></div>
<p>Questions:<br>
1.1 What is the Fraction of Variance for PC1?<br>
1.2 How many differentially expressed genes are there for treated vs untreated (FDR &lt; 0.05)? 1.3 What is the FBgn id of the 10th most differentially expressed gene (by smallest P value)?</p>
<p>Extra 1.4 What code can generate a heatmap of variable genes (starting from <code>count_scaled</code>)? 1.5 What code can you use to visualise expression of the pasilla gene (gene id: FBgn0261552)<br>
1.6 What code can generate an interactive volcano plot that has gene ids showing on hover?<br>
1.7 What code can generate a heatmap of the top 100 differentially expressed genes?</p>
</div>
</div>
<div id="part-2-bulk-rna-seq-extended" class="section level1">
<h1 class="hasAnchor">
<a href="#part-2-bulk-rna-seq-extended" class="anchor"></a>Part 2 Bulk RNA-seq Extended</h1>
<p><em>tidybulk</em> integrates several popular methods for differential transcript abundance testing: the edgeR quasi-likelihood <span class="citation">(Chen, Lun, and Smyth 2016)</span> (tidybulk default method), edgeR likelihood ratio <span class="citation">(McCarthy, Chen, and Smyth 2012)</span>, limma-voom <span class="citation">(Law et al. 2014)</span> and DESeq2 <span class="citation">(Love, Huber, and Anders 2014)</span>. A common question researchers have is which method to choose. Mike Love, DESeq2 author has this advice in his <a href="https://mikelove.wordpress.com/2016/09/28/deseq2-or-edger">blog</a>.</p>
<p><img src="../inst/vignettes/which_method.png" width="350px"></p>
<p>tidybulk can help you decide which method (or methods) to use, as it provides an easy way to run multiple and see how they compare.</p>
<div id="comparison-of-differential-analysis-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#comparison-of-differential-analysis-methods" class="anchor"></a>Comparison of differential analysis methods</h2>
<p>We can perform differential analysis with several methods, and the results will be added to the original dataset.</p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="co"># load additional libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://forcats.tidyverse.org">forcats</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidygate</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally">GGally</a></span><span class="op">)</span></pre></div>
<p>As before, we first pre-process the data, creating a tibble and identifying abundant genes.</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="va">pasilla_de</span> <span class="op">&lt;-</span> 
  <span class="fu">rpharma2020tidytranscriptomics</span><span class="fu">::</span><span class="va"><a href="../reference/pasilla.html">pasilla</a></span> <span class="op">%&gt;%</span> 
  
  <span class="co"># Convert SummarizedExperiment object to tibble</span>
  <span class="va">tidybulk</span> <span class="op">%&gt;%</span>
    
  <span class="co"># Add gene symbols</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Dm.eg.db</span>,   keys<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span>, keytype <span class="op">=</span> <span class="st">"FLYBASE"</span>, column<span class="op">=</span><span class="st">"SYMBOL"</span>, multiVals <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Scale abundance</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span><span class="op">(</span>factor_of_interest<span class="op">=</span><span class="va">condition</span><span class="op">)</span> 
<span class="co">#&gt; 'select()' returned 1:1 mapping between keys and columns</span></pre></div>
<p>This is an example for the default method for differential abundance testing. It uses the edgeR quasi-likelihood method.</p>
<div class="sourceCode" id="cb35"><pre class="downlit">
<span class="va">pasilla_de</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Test differential composition</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">condition</span> <span class="op">+</span> <span class="va">type</span>, 
    action<span class="op">=</span><span class="st">"get"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Sort by P value        </span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">PValue</span><span class="op">)</span>
<span class="co">#&gt; tidybulk says: All methods use raw counts, </span>
<span class="co">#&gt; irrespective of if scale_abundance or adjust_abundance have been calculated, </span>
<span class="co">#&gt; therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</span>
<span class="co">#&gt; tidybulk says: The design column names are "(Intercept), conditionuntreated, typesingle_end"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span>
<span class="co">#&gt; # A tibble: 7,919 x 8</span>
<span class="co">#&gt;    feature     symbol .abundant logFC logCPM     F   PValue          FDR</span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;  &lt;lgl&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt;  1 FBgn0025111 Ant2   TRUE      -2.86   6.92 1304. 2.37e-12 0.0000000149</span>
<span class="co">#&gt;  2 FBgn0039155 Kal1   TRUE       4.61   5.88 1194. 3.75e-12 0.0000000149</span>
<span class="co">#&gt;  3 FBgn0003360 sesB   TRUE       3.12   8.45  792. 3.17e-11 0.0000000837</span>
<span class="co">#&gt;  4 FBgn0035085 CG3770 TRUE       2.57   5.68  728. 4.89e-11 0.0000000968</span>
<span class="co">#&gt;  5 FBgn0039827 CG1544 TRUE       4.17   4.39  616. 1.16e-10 0.000000169 </span>
<span class="co">#&gt;  6 FBgn0029167 Hml    TRUE       2.19   8.22  599. 1.34e-10 0.000000169 </span>
<span class="co">#&gt;  7 FBgn0034736 gas    TRUE       3.52   4.18  586. 1.50e-10 0.000000169 </span>
<span class="co">#&gt;  8 FBgn0026562 SPARC  TRUE       2.47  11.8   504. 3.27e-10 0.000000323 </span>
<span class="co">#&gt;  9 FBgn0034434 &lt;NA&gt;   TRUE       3.63   3.21  390. 1.22e- 9 0.000000983 </span>
<span class="co">#&gt; 10 FBgn0000071 Ama    TRUE      -2.63   4.79  389. 1.24e- 9 0.000000983 </span>
<span class="co">#&gt; # … with 7,909 more rows</span></pre></div>
<p>Now let’s try to perform multiple methods on the same dataset.</p>
<div class="sourceCode" id="cb36"><pre class="downlit">
<span class="va">de_all</span> <span class="op">&lt;-</span> 
  
  <span class="va">pasilla_de</span> <span class="op">%&gt;%</span>
  
  <span class="co"># edgeR QLT</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">condition</span> <span class="op">+</span> <span class="va">type</span>, 
    method <span class="op">=</span> <span class="st">"edger_quasi_likelihood"</span>,
    prefix <span class="op">=</span> <span class="st">"edgerQLT_"</span>
  <span class="op">)</span>  <span class="op">%&gt;%</span>
  
  <span class="co"># edgeR LRT</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">condition</span> <span class="op">+</span> <span class="va">type</span>, 
    method <span class="op">=</span> <span class="st">"edger_likelihood_ratio"</span>,
    prefix <span class="op">=</span> <span class="st">"edgerLR_"</span>
  <span class="op">)</span>  <span class="op">%&gt;%</span>
  
  <span class="co"># limma-voom</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">condition</span> <span class="op">+</span> <span class="va">type</span>, 
    method <span class="op">=</span> <span class="st">"limma_voom"</span>,
    prefix <span class="op">=</span> <span class="st">"voom_"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># DESeq2</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">condition</span> <span class="op">+</span> <span class="va">type</span>, 
    method <span class="op">=</span> <span class="st">"deseq2"</span>,
    prefix <span class="op">=</span> <span class="st">"deseq2_"</span>
  <span class="op">)</span> 

<span class="co"># take a look</span>

<span class="va">de_all</span>
<span class="co">#&gt; # A tibble: 55,433 x 29</span>
<span class="co">#&gt;    feature sample counts condition type  symbol .abundant edgerQLT_logFC</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;fct&gt;   &lt;int&gt; &lt;fct&gt;     &lt;fct&gt; &lt;chr&gt;  &lt;lgl&gt;              &lt;dbl&gt;</span>
<span class="co">#&gt;  1 FBgn00… untrt1     92 untreated sing… a      TRUE              0.0337</span>
<span class="co">#&gt;  2 FBgn00… untrt1   4664 untreated sing… Abl    TRUE              0.249 </span>
<span class="co">#&gt;  3 FBgn00… untrt1    583 untreated sing… abo    TRUE              0.0569</span>
<span class="co">#&gt;  4 FBgn00… untrt1   1446 untreated sing… Acph-1 TRUE              0.0402</span>
<span class="co">#&gt;  5 FBgn00… untrt1     15 untreated sing… mAChR… TRUE             -0.387 </span>
<span class="co">#&gt;  6 FBgn00… untrt1 101664 untreated sing… Act5C  TRUE             -0.310 </span>
<span class="co">#&gt;  7 FBgn00… untrt1  33402 untreated sing… Act42A TRUE             -0.617 </span>
<span class="co">#&gt;  8 FBgn00… untrt1     21 untreated sing… Act57B TRUE             -0.491 </span>
<span class="co">#&gt;  9 FBgn00… untrt1     12 untreated sing… Act87E TRUE             -0.610 </span>
<span class="co">#&gt; 10 FBgn00… untrt1   3026 untreated sing… Pfas   TRUE              0.163 </span>
<span class="co">#&gt; # … with 55,423 more rows, and 21 more variables: edgerQLT_logCPM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerQLT_F &lt;dbl&gt;, edgerQLT_PValue &lt;dbl&gt;, edgerQLT_FDR &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerLR_logFC &lt;dbl&gt;, edgerLR_logCPM &lt;dbl&gt;, edgerLR_LR &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerLR_PValue &lt;dbl&gt;, edgerLR_FDR &lt;dbl&gt;, voom_logFC &lt;dbl&gt;,</span>
<span class="co">#&gt; #   voom_AveExpr &lt;dbl&gt;, voom_t &lt;dbl&gt;, voom_P.Value &lt;dbl&gt;, voom_adj.P.Val &lt;dbl&gt;,</span>
<span class="co">#&gt; #   voom_B &lt;dbl&gt;, deseq2_baseMean &lt;dbl&gt;, deseq2_log2FoldChange &lt;dbl&gt;,</span>
<span class="co">#&gt; #   deseq2_lfcSE &lt;dbl&gt;, deseq2_stat &lt;dbl&gt;, deseq2_pvalue &lt;dbl&gt;,</span>
<span class="co">#&gt; #   deseq2_padj &lt;dbl&gt;</span></pre></div>
<hr>
<div id="note" class="section level4">
<h4 class="hasAnchor">
<a href="#note" class="anchor"></a>Note</h4>
<p>You may notice that the methods produce columns with different names for similar outputs, for example edgerQLT_PValue vs deseq2_pvalue and edgerQLT_FDR vs deseq2_padj. If you wish to make these consistent you can do that with tidyverse <code>rename</code>.</p>
<div class="sourceCode" id="cb37"><pre class="downlit">
<span class="co"># example of renaming DESeq2 columns</span>
<span class="va">de_all</span> <span class="op">%&gt;%</span>
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>deseq2_PValue <span class="op">=</span> <span class="va">deseq2_pvalue</span>, deseq2_FDR <span class="op">=</span> <span class="va">deseq2_padj</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 55,433 x 29</span>
<span class="co">#&gt;    feature sample counts condition type  symbol .abundant edgerQLT_logFC</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;fct&gt;   &lt;int&gt; &lt;fct&gt;     &lt;fct&gt; &lt;chr&gt;  &lt;lgl&gt;              &lt;dbl&gt;</span>
<span class="co">#&gt;  1 FBgn00… untrt1     92 untreated sing… a      TRUE              0.0337</span>
<span class="co">#&gt;  2 FBgn00… untrt1   4664 untreated sing… Abl    TRUE              0.249 </span>
<span class="co">#&gt;  3 FBgn00… untrt1    583 untreated sing… abo    TRUE              0.0569</span>
<span class="co">#&gt;  4 FBgn00… untrt1   1446 untreated sing… Acph-1 TRUE              0.0402</span>
<span class="co">#&gt;  5 FBgn00… untrt1     15 untreated sing… mAChR… TRUE             -0.387 </span>
<span class="co">#&gt;  6 FBgn00… untrt1 101664 untreated sing… Act5C  TRUE             -0.310 </span>
<span class="co">#&gt;  7 FBgn00… untrt1  33402 untreated sing… Act42A TRUE             -0.617 </span>
<span class="co">#&gt;  8 FBgn00… untrt1     21 untreated sing… Act57B TRUE             -0.491 </span>
<span class="co">#&gt;  9 FBgn00… untrt1     12 untreated sing… Act87E TRUE             -0.610 </span>
<span class="co">#&gt; 10 FBgn00… untrt1   3026 untreated sing… Pfas   TRUE              0.163 </span>
<span class="co">#&gt; # … with 55,423 more rows, and 21 more variables: edgerQLT_logCPM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerQLT_F &lt;dbl&gt;, edgerQLT_PValue &lt;dbl&gt;, edgerQLT_FDR &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerLR_logFC &lt;dbl&gt;, edgerLR_logCPM &lt;dbl&gt;, edgerLR_LR &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerLR_PValue &lt;dbl&gt;, edgerLR_FDR &lt;dbl&gt;, voom_logFC &lt;dbl&gt;,</span>
<span class="co">#&gt; #   voom_AveExpr &lt;dbl&gt;, voom_t &lt;dbl&gt;, voom_P.Value &lt;dbl&gt;, voom_adj.P.Val &lt;dbl&gt;,</span>
<span class="co">#&gt; #   voom_B &lt;dbl&gt;, deseq2_baseMean &lt;dbl&gt;, deseq2_log2FoldChange &lt;dbl&gt;,</span>
<span class="co">#&gt; #   deseq2_lfcSE &lt;dbl&gt;, deseq2_stat &lt;dbl&gt;, deseq2_PValue &lt;dbl&gt;,</span>
<span class="co">#&gt; #   deseq2_FDR &lt;dbl&gt;</span></pre></div>
<hr>
<p>We can visually compare the log fold change (logFC) of transcript abundance for the comparison of interest (treated vs untreated) for all methods. We will notice that the consistency of the logFC is really high for the methods.</p>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="va">de_all</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">edgerQLT_logFC</span>, <span class="va">edgerLR_logFC</span>, <span class="va">voom_logFC</span>, <span class="va">deseq2_log2FoldChange</span>, <span class="va">feature</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-41-1.png" width="700"></p>
<p>Similarly, we can visually compare the significance for all methods. In this case the difference is larger.</p>
<div class="sourceCode" id="cb39"><pre class="downlit">
<span class="va">de_all</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">edgerQLT_PValue</span>, <span class="va">edgerLR_PValue</span>, <span class="va">voom_P.Value</span>, <span class="va">deseq2_pvalue</span>, <span class="va">feature</span> <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-42-1.png" width="700"></p>
<p>We can select some of the transcripts for further analysis using the <a href="https://github.com/stemangiola/tidygate"><em>tidygate</em></a> package.</p>
<div class="sourceCode" id="cb40"><pre class="downlit">
<span class="va">de_gate</span> <span class="op">&lt;-</span> 
  <span class="va">de_all</span> <span class="op">%&gt;%</span>
  
  <span class="fu"><a href="https://rdrr.io/pkg/tidygate/man/gate-methods.html">gate</a></span><span class="op">(</span>
    <span class="va">feature</span>,
    <span class="va">edgerQLT_PValue</span>, 
    <span class="va">deseq2_pvalue</span>, 
    opacity<span class="op">=</span><span class="fl">0.3</span>, 
    how_many_gates <span class="op">=</span> <span class="fl">2</span> 
  <span class="op">)</span></pre></div>
<pre><code>#&gt; # A tibble: 55,433 x 30
#&gt;    feature sample counts condition type  symbol .abundant edgerQLT_logFC
#&gt;    &lt;chr&gt;   &lt;fct&gt;   &lt;int&gt; &lt;fct&gt;     &lt;fct&gt; &lt;chr&gt;  &lt;lgl&gt;              &lt;dbl&gt;
#&gt;  1 FBgn00… untrt1     92 untreated sing… a      TRUE              0.0337
#&gt;  2 FBgn00… untrt1   4664 untreated sing… Abl    TRUE              0.249 
#&gt;  3 FBgn00… untrt1    583 untreated sing… abo    TRUE              0.0569
#&gt;  4 FBgn00… untrt1   1446 untreated sing… Acph-1 TRUE              0.0402
#&gt;  5 FBgn00… untrt1     15 untreated sing… mAChR… TRUE             -0.387 
#&gt;  6 FBgn00… untrt1 101664 untreated sing… Act5C  TRUE             -0.310 
#&gt;  7 FBgn00… untrt1  33402 untreated sing… Act42A TRUE             -0.617 
#&gt;  8 FBgn00… untrt1     21 untreated sing… Act57B TRUE             -0.491 
#&gt;  9 FBgn00… untrt1     12 untreated sing… Act87E TRUE             -0.610 
#&gt; 10 FBgn00… untrt1   3026 untreated sing… Pfas   TRUE              0.163 
#&gt; # … with 55,423 more rows, and 22 more variables: edgerQLT_logCPM &lt;dbl&gt;,
#&gt; #   edgerQLT_F &lt;dbl&gt;, edgerQLT_PValue &lt;dbl&gt;, edgerQLT_FDR &lt;dbl&gt;,
#&gt; #   edgerLR_logFC &lt;dbl&gt;, edgerLR_logCPM &lt;dbl&gt;, edgerLR_LR &lt;dbl&gt;,
#&gt; #   edgerLR_PValue &lt;dbl&gt;, edgerLR_FDR &lt;dbl&gt;, voom_logFC &lt;dbl&gt;,
#&gt; #   voom_AveExpr &lt;dbl&gt;, voom_t &lt;dbl&gt;, voom_P.Value &lt;dbl&gt;, voom_adj.P.Val &lt;dbl&gt;,
#&gt; #   voom_B &lt;dbl&gt;, deseq2_baseMean &lt;dbl&gt;, deseq2_log2FoldChange &lt;dbl&gt;,
#&gt; #   deseq2_lfcSE &lt;dbl&gt;, deseq2_stat &lt;dbl&gt;, deseq2_pvalue &lt;dbl&gt;,
#&gt; #   deseq2_padj &lt;dbl&gt;, gate &lt;chr&gt;</code></pre>
<p><img src="../inst/vignettes/comparison_different_DE_methods_gates.png" width="322"></p>
<p>We can now select the transcripts from the two gates (i.e. oversignificant in DESeq2 and oversignificant in edgeR)</p>
<div class="sourceCode" id="cb42"><pre class="downlit">
<span class="va">de_gate</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Filter only transcripts within the gates</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">gate</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  
  <span class="co"># Rename for clarity</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>gate <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span>
    <span class="va">gate</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">~</span> <span class="st">"more in edgeR"</span>,
    <span class="va">gate</span> <span class="op">==</span> <span class="fl">2</span> <span class="op">~</span> <span class="st">"more in DESeq2"</span>,
    <span class="cn">TRUE</span> <span class="op">~</span> <span class="va">gate</span>
  <span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Rearrange order</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_reorder.html">fct_reorder</a></span><span class="op">(</span><span class="va">feature</span>, <span class="va">edgerQLT_PValue</span>, <span class="va">min</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Plot</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">condition</span>, <span class="va">counts_scaled</span>, color<span class="op">=</span><span class="va">gate</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">feature</span>, scale<span class="op">=</span><span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-46-1.png" width="700"></p>
<p>This enables us to see, for example, that DESeq2 produces a more conservative statistic for the transcript <code>FBgn0052939</code> .</p>
<div class="sourceCode" id="cb43"><pre class="downlit">
<span class="va">de_gate</span> <span class="op">%&gt;%</span>
  <span class="va">pivot_transcript</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">==</span> <span class="st">"FBgn0052939"</span><span class="op">)</span><span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">edgerQLT_logFC</span>, <span class="va">deseq2_log2FoldChange</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1 x 2</span>
<span class="co">#&gt;   edgerQLT_logFC deseq2_log2FoldChange</span>
<span class="co">#&gt;            &lt;dbl&gt;                 &lt;dbl&gt;</span>
<span class="co">#&gt; 1           2.37                  1.65</span></pre></div>
</div>
</div>
<div id="study-of-cellularity" class="section level2">
<h2 class="hasAnchor">
<a href="#study-of-cellularity" class="anchor"></a>Study of cellularity</h2>
<p>If we are sequencing tissue samples, we may want to know what cell types are present, and if there are differences in expression between them. <code>tidybulk</code> has a <code>deconvolve_cellularity</code> function that can help us do this.</p>
<p>For this example we will use a subset of the breast cancer TCGA dataset.</p>
<div class="sourceCode" id="cb44"><pre class="downlit">

<span class="va">BRCA_tidy</span> <span class="op">=</span> 
    <span class="fu">rpharma2020tidytranscriptomics</span><span class="fu">::</span><span class="va"><a href="../reference/BRCA.html">BRCA</a></span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="va">patient</span>, <span class="va">transcript</span>, <span class="va">count</span><span class="op">)</span>

<span class="va">BRCA_tidy</span>
<span class="co">#&gt; # A tibble: 198,374 x 5</span>
<span class="co">#&gt;    patient      transcript count PFI.time.2 PFI.2</span>
<span class="co">#&gt;    &lt;fct&gt;        &lt;fct&gt;      &lt;int&gt;      &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt;  1 TCGA-C8-A275 C1orf112    1571          1     0</span>
<span class="co">#&gt;  2 TCGA-C8-A275 FGR         1873          1     0</span>
<span class="co">#&gt;  3 TCGA-C8-A275 FUCA2       3770          1     0</span>
<span class="co">#&gt;  4 TCGA-C8-A275 GCLC        2738          1     0</span>
<span class="co">#&gt;  5 TCGA-C8-A275 LAS1L       3175          1     0</span>
<span class="co">#&gt;  6 TCGA-C8-A275 ENPP4       1340          1     0</span>
<span class="co">#&gt;  7 TCGA-C8-A275 SEMA3F      5479          1     0</span>
<span class="co">#&gt;  8 TCGA-C8-A275 RAD52        472          1     0</span>
<span class="co">#&gt;  9 TCGA-C8-A275 BAD         1761          1     0</span>
<span class="co">#&gt; 10 TCGA-C8-A275 CD99       14697          1     0</span>
<span class="co">#&gt; # … with 198,364 more rows</span></pre></div>
<p>We can easily infer tissue cellularity across a range of methods (Cibersort, EPIC and ls_fit) with LM22 or custom signature</p>
<div class="sourceCode" id="cb45"><pre class="downlit">
<span class="va">BRCA_cell_type</span> <span class="op">=</span> 
    <span class="va">BRCA_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>action<span class="op">=</span><span class="st">"get"</span><span class="op">)</span>
<span class="co">#&gt; Installing e1071 needed for Cibersort</span>
<span class="co">#&gt; Installing package into '/tmp/RtmpDN9wdy/temp_libpath1f114ac21f82'</span>
<span class="co">#&gt; (as 'lib' is unspecified)</span>

<span class="va">BRCA_cell_type</span>
<span class="co">#&gt; # A tibble: 22 x 25</span>
<span class="co">#&gt;    patient PFI.time.2 PFI.2 `cibersort: B c… `cibersort: B c… `cibersort: Pla…</span>
<span class="co">#&gt;    &lt;chr&gt;        &lt;dbl&gt; &lt;int&gt;            &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;</span>
<span class="co">#&gt;  1 TCGA-C…          1     0          0.0325            0               0.0127 </span>
<span class="co">#&gt;  2 TCGA-A…          1     0          0.0170            0               0      </span>
<span class="co">#&gt;  3 TCGA-A…          5     0          0.00127           0               0      </span>
<span class="co">#&gt;  4 TCGA-C…          5     0          0.0372            0               0.0228 </span>
<span class="co">#&gt;  5 TCGA-P…          5     0          0.00453           0               0.0260 </span>
<span class="co">#&gt;  6 TCGA-B…       6292     0          0.144             0               0      </span>
<span class="co">#&gt;  7 TCGA-B…       7106     0          0.0632            0               0.0162 </span>
<span class="co">#&gt;  8 TCGA-B…       7777     0          0                 0.0142          0.00315</span>
<span class="co">#&gt;  9 TCGA-B…       8008     0          0.203             0               0.0228 </span>
<span class="co">#&gt; 10 TCGA-B…       8391     0          0.147             0               0      </span>
<span class="co">#&gt; # … with 12 more rows, and 19 more variables: `cibersort: T cells CD8` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `cibersort: T cells CD4 naive` &lt;dbl&gt;, `cibersort: T cells CD4 memory</span>
<span class="co">#&gt; #   resting` &lt;dbl&gt;, `cibersort: T cells CD4 memory activated` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `cibersort: T cells follicular helper` &lt;dbl&gt;, `cibersort: T cells</span>
<span class="co">#&gt; #   regulatory (Tregs)` &lt;dbl&gt;, `cibersort: T cells gamma delta` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `cibersort: NK cells resting` &lt;dbl&gt;, `cibersort: NK cells activated` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `cibersort: Monocytes` &lt;dbl&gt;, `cibersort: Macrophages M0` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `cibersort: Macrophages M1` &lt;dbl&gt;, `cibersort: Macrophages M2` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `cibersort: Dendritic cells resting` &lt;dbl&gt;, `cibersort: Dendritic cells</span>
<span class="co">#&gt; #   activated` &lt;dbl&gt;, `cibersort: Mast cells resting` &lt;dbl&gt;, `cibersort: Mast</span>
<span class="co">#&gt; #   cells activated` &lt;dbl&gt;, `cibersort: Eosinophils` &lt;dbl&gt;, `cibersort:</span>
<span class="co">#&gt; #   Neutrophils` &lt;dbl&gt;</span></pre></div>
<p>Cell type proportions are added to the tibble as new columns. The prefix makes it easy to reshape the data frame if needed, for visualisation or further analyses.</p>
<div class="sourceCode" id="cb46"><pre class="downlit">
<span class="va">BRCA_cell_type_long</span> <span class="op">=</span> 
    <span class="va">BRCA_cell_type</span> <span class="op">%&gt;%</span>
    
    <span class="co"># Reshape</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>
        <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span><span class="op">(</span><span class="st">"cibersort"</span><span class="op">)</span>, 
        names_prefix <span class="op">=</span> <span class="st">"cibersort: "</span>, 
        names_to <span class="op">=</span> <span class="st">"cell_type"</span>,
        values_to <span class="op">=</span> <span class="st">"proportion"</span>
    <span class="op">)</span> 

<span class="va">BRCA_cell_type_long</span>
<span class="co">#&gt; # A tibble: 484 x 5</span>
<span class="co">#&gt;    patient      PFI.time.2 PFI.2 cell_type                    proportion</span>
<span class="co">#&gt;    &lt;chr&gt;             &lt;dbl&gt; &lt;int&gt; &lt;chr&gt;                             &lt;dbl&gt;</span>
<span class="co">#&gt;  1 TCGA-C8-A275          1     0 B cells naive                    0.0325</span>
<span class="co">#&gt;  2 TCGA-C8-A275          1     0 B cells memory                   0     </span>
<span class="co">#&gt;  3 TCGA-C8-A275          1     0 Plasma cells                     0.0127</span>
<span class="co">#&gt;  4 TCGA-C8-A275          1     0 T cells CD8                      0     </span>
<span class="co">#&gt;  5 TCGA-C8-A275          1     0 T cells CD4 naive                0.0507</span>
<span class="co">#&gt;  6 TCGA-C8-A275          1     0 T cells CD4 memory resting       0.0299</span>
<span class="co">#&gt;  7 TCGA-C8-A275          1     0 T cells CD4 memory activated     0.0332</span>
<span class="co">#&gt;  8 TCGA-C8-A275          1     0 T cells follicular helper        0     </span>
<span class="co">#&gt;  9 TCGA-C8-A275          1     0 T cells regulatory (Tregs)       0.0326</span>
<span class="co">#&gt; 10 TCGA-C8-A275          1     0 T cells gamma delta              0.0247</span>
<span class="co">#&gt; # … with 474 more rows</span></pre></div>
<p>Note: Variable creation is functional to education, however is not needed nor recommended in normal circumstances.</p>
<div class="sourceCode" id="cb47"><pre class="downlit">
<span class="va">BRCA_cell_type_long</span> <span class="op">%&gt;%</span>
    
    <span class="co"># Plot</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">patient</span>, y<span class="op">=</span><span class="va">proportion</span>, fill<span class="op">=</span><span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span> </pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-51-1.png" width="700"></p>
<p>We can visualise the tissue composition similarity performing a dimensionality reduction on cell type and proportion (rather than on transcript and counts)</p>
<div class="sourceCode" id="cb48"><pre class="downlit">

<span class="va">BRCA_cell_type_long</span> <span class="op">%&gt;%</span>
    
    <span class="co"># Add some noise to the proportion inference to avoid PCA error</span>
    <span class="co"># A better (but more complex) way would be to eliminate the zero variance cell types </span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>proportion <span class="op">=</span> <span class="va">proportion</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fl">1e-10</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">abs</span><span class="op">)</span> <span class="op">%&gt;%</span>
    
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>
        <span class="va">patient</span>, 
        <span class="va">cell_type</span>, 
        <span class="va">proportion</span>, 
        method<span class="op">=</span><span class="st">"PCA"</span>, 
        action<span class="op">=</span><span class="st">"get"</span>
    <span class="op">)</span> <span class="op">%&gt;%</span>
    
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">PC1</span>, <span class="va">PC2</span>, label<span class="op">=</span><span class="va">patient</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"red"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggrepel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
    <span class="va">custom_theme</span>
<span class="co">#&gt; Warning in eval(dots[[i]][[action]], env, env): tidybulk says: highly abundant</span>
<span class="co">#&gt; transcripts were not identified (i.e. identify_abundant()) or filtered (i.e.,</span>
<span class="co">#&gt; keep_abundant), therefore this operation will be performed on unfiltered</span>
<span class="co">#&gt; data. In rare occasions this could be wanted. In standard whole-transcriptome</span>
<span class="co">#&gt; workflows is generally unwanted.</span>
<span class="co">#&gt; Getting the 22 most variable genes</span>
<span class="co">#&gt; Warning in as_mapper(.f2)(.x): </span>
<span class="co">#&gt;                      tidybulk says: In PCA correlation there is &lt; 100 genes that have non NA values is all samples.</span>
<span class="co">#&gt; The correlation calculation would not be reliable,</span>
<span class="co">#&gt; we suggest to partition the dataset for sample clusters.</span>
<span class="co">#&gt;                      </span>
<span class="co">#&gt; Fraction of variance explained by the selected principal components</span>
<span class="co">#&gt; # A tibble: 2 x 2</span>
<span class="co">#&gt;   `Fraction of variance`    PC</span>
<span class="co">#&gt;                    &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1                  0.175     1</span>
<span class="co">#&gt; 2                  0.146     2</span>
<span class="co">#&gt; tidybulk says: to access the raw results do `attr(..., "internals")$PCA`</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-52-1.png" width="700"></p>
<p>We can also perform differential tissue composition analyses, similarly to how we performed differential transcript abundance analyses. We can perform our analyses on censored (survival analyses) or known factor of interest.</p>
<div class="sourceCode" id="cb49"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival">survival</a></span><span class="op">)</span>

<span class="va">BRCA_tidy_survival</span> <span class="op">=</span> 
    <span class="va">BRCA_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_cellularity-methods.html">test_differential_cellularity</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html">Surv</a></span><span class="op">(</span><span class="va">PFI.time.2</span>, <span class="va">PFI.2</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>   <span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span>
<span class="co">#&gt; Installing broom needed for analyses</span>
<span class="co">#&gt; Installing package into '/tmp/RtmpDN9wdy/temp_libpath1f114ac21f82'</span>
<span class="co">#&gt; (as 'lib' is unspecified)</span>

<span class="va">BRCA_tidy_survival</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">.cell_type</span>, <span class="va">p.value</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 22 x 6</span>
<span class="co">#&gt;    .cell_type            p.value cell_type_proport… estimate std.error statistic</span>
<span class="co">#&gt;    &lt;chr&gt;                   &lt;dbl&gt; &lt;list&gt;                &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 NK cells resting       0.0863 &lt;tibble [22 × 6]&gt;     0.330     0.192     1.72 </span>
<span class="co">#&gt;  2 T cells CD4 memory a…  0.0884 &lt;tibble [22 × 6]&gt;     0.313     0.184     1.70 </span>
<span class="co">#&gt;  3 Mast cells resting     0.150  &lt;tibble [22 × 6]&gt;    -0.379     0.264    -1.44 </span>
<span class="co">#&gt;  4 T cells CD4 memory r…  0.161  &lt;tibble [22 × 6]&gt;    -0.865     0.617    -1.40 </span>
<span class="co">#&gt;  5 Macrophages M2         0.164  &lt;tibble [22 × 6]&gt;     0.365     0.262     1.39 </span>
<span class="co">#&gt;  6 T cells gamma delta    0.291  &lt;tibble [22 × 6]&gt;    -0.534     0.506    -1.06 </span>
<span class="co">#&gt;  7 T cells regulatory (…  0.330  &lt;tibble [22 × 6]&gt;    -0.496     0.509    -0.975</span>
<span class="co">#&gt;  8 T cells follicular h…  0.372  &lt;tibble [22 × 6]&gt;    -0.205     0.229    -0.893</span>
<span class="co">#&gt;  9 Macrophages M0         0.520  &lt;tibble [22 × 6]&gt;    -0.248     0.385    -0.643</span>
<span class="co">#&gt; 10 T cells CD4 naive      0.579  &lt;tibble [22 × 6]&gt;     0.113     0.204     0.555</span>
<span class="co">#&gt; # … with 12 more rows</span></pre></div>
<p>We can visualise the cell type proportions for the most associated cell types</p>
<div class="sourceCode" id="cb50"><pre class="downlit">
<span class="va">BRCA_tidy_survival</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">cell_type_proportions</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">PFI.time.2</span>, <span class="va">.proportion</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">PFI.2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">.cell_type</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"logit"</span><span class="op">)</span> <span class="op">+</span>
    <span class="va">custom_theme</span>
<span class="co">#&gt; Warning: Transformation introduced infinite values in continuous y-axis</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-54-1.png" width="700"></p>
</div>
<div id="nested-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#nested-analyses" class="anchor"></a>Nested analyses</h2>
<p><code>tidybulk</code> allows for data nesting, using the <code>tidyr</code> utility <code>nest</code>. This is an extremely powerful tool as allows functional programming on transcriptomic data.</p>
<div id="perform-same-analysis-on-subsets" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-same-analysis-on-subsets" class="anchor"></a>Perform same analysis on subsets</h3>
<p>Let’s suppose we want to perform differential transcript abundance analysis independently for two different data subsets to compare results after the test</p>
<div class="sourceCode" id="cb51"><pre class="downlit">
<span class="co"># load additional libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb52"><pre class="downlit">
<span class="va">pasilla_de</span> <span class="op">&lt;-</span> 
  <span class="fu">rpharma2020tidytranscriptomics</span><span class="fu">::</span><span class="va"><a href="../reference/pasilla.html">pasilla</a></span> <span class="op">%&gt;%</span> 
  
  <span class="co"># Convert SummarizedExperiment object to tibble</span>
  <span class="va">tidybulk</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Filter counts</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest<span class="op">=</span><span class="va">condition</span><span class="op">)</span> <span class="op">%&gt;%</span>
    
  <span class="co"># Scale abundance   </span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Nest</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">type</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Differential analysis</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span>
    <span class="va">data</span>,
    <span class="op">~</span>  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span><span class="va">.x</span>, <span class="op">~</span> <span class="va">condition</span><span class="op">)</span> 
  <span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>
<span class="co">#&gt; tidybulk says: All methods use raw counts, </span>
<span class="co">#&gt; irrespective of if scale_abundance or adjust_abundance have been calculated, </span>
<span class="co">#&gt; therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</span>
<span class="co">#&gt; tidybulk says: The design column names are "(Intercept), conditionuntreated"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span>
<span class="co">#&gt; tidybulk says: All methods use raw counts, </span>
<span class="co">#&gt; irrespective of if scale_abundance or adjust_abundance have been calculated, </span>
<span class="co">#&gt; therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</span>
<span class="co">#&gt; tidybulk says: The design column names are "(Intercept), conditionuntreated"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span></pre></div>
<p>Now we can for example compare the number of differentially transcribed genes and their co-expression</p>
<div class="sourceCode" id="cb53"><pre class="downlit">
<span class="va">pasilla_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">type</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>
    number_of_differential <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_int</a></span><span class="op">(</span>
      <span class="va">data</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">nrow</span>
    <span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 2 x 3</span>
<span class="co">#&gt;   type       data                   number_of_differential</span>
<span class="co">#&gt;   &lt;fct&gt;      &lt;list&gt;                                  &lt;int&gt;</span>
<span class="co">#&gt; 1 single_end &lt;tibble [43,797 × 13]&gt;                     40</span>
<span class="co">#&gt; 2 paired_end &lt;tibble [58,396 × 13]&gt;                   1270</span></pre></div>
<p>We can easily see which genes overlap, and plot them</p>
<div class="sourceCode" id="cb54"><pre class="downlit">
<span class="va">pasilla_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">feature</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>occurrences <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_int</a></span><span class="op">(</span><span class="va">data</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span><span class="op">(</span><span class="va">type</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">nrow</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># We filter some of them</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">occurrences</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># And plot</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">type</span>, <span class="va">counts_scaled</span> <span class="op">+</span><span class="fl">1</span>, color<span class="op">=</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-58-1.png" width="700"></p>
</div>
</div>
<div id="key-points-1" class="section level2">
<h2 class="hasAnchor">
<a href="#key-points-1" class="anchor"></a>Key Points</h2>
<ul>
<li>
<code>tidybulk</code> allows streamlined multi-method analyses</li>
<li>
<code>tidygate</code> allows the selection of arbitrary points in a two-dimensional plot, and add the gate information to the input tibble</li>
<li>
<code>tidybulk</code> allow easy deconvolution analyses</li>
<li>Hypothesis testing for tissue composition is analogous to the one for transcript abundance</li>
<li>Nesting allows for iteratively perform analyses on portions of the data set, without the creation of new variable, and in a totally functional way.</li>
</ul>
</div>
<div id="exercises-1" class="section level2">
<h2 class="hasAnchor">
<a href="#exercises-1" class="anchor"></a>Exercises</h2>
<p>Questions: 2.1 How can I compare the amount of differential abundant transcripts for the different methods? 2.2 How can I trace the adjusted p-value rank for methods?</p>
</div>
</div>
<div id="part-3-single-cell-rna-seq" class="section level1">
<h1 class="hasAnchor">
<a href="#part-3-single-cell-rna-seq" class="anchor"></a>Part 3 Single-cell RNA-seq</h1>
<p>tidyseurat provides a bridge between the Seurat single-cell package <span class="citation">(Butler et al. 2018; Stuart et al. 2019)</span> and the tidyverse <span class="citation">(Wickham et al. 2019)</span>. It creates an invisible layer that enables viewing the Seurat object as a tidyverse tibble, and provides Seurat-compatible <em>dplyr</em>, <em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions.</p>
<div class="sourceCode" id="cb55"><pre class="downlit">
<span class="co"># load additional libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidyseurat</span><span class="op">)</span></pre></div>
<div id="create-tidyseurat" class="section level2">
<h2 class="hasAnchor">
<a href="#create-tidyseurat" class="anchor"></a>Create <code>tidyseurat</code>
</h2>
<p>This is a seurat object but it is evaluated as tibble. So it is fully compatible both with Seurat and tidyverse APIs.</p>
<div class="sourceCode" id="cb56"><pre class="downlit">
<span class="va">pbmc_small_tidy</span> <span class="op">&lt;-</span> <span class="fu">tidyseurat</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/tidyseurat/man/pbmc_small.html">pbmc_small</a></span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb57"><pre class="downlit">
<span class="va">pbmc_small_tidy</span>
<span class="co">#&gt; # A tibble: 80 x 16</span>
<span class="co">#&gt;    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span>
<span class="co">#&gt;  1 ATGC… SeuratPro…         70           47 0               A             g2    </span>
<span class="co">#&gt;  2 CATG… SeuratPro…         85           52 0               A             g1    </span>
<span class="co">#&gt;  3 GAAC… SeuratPro…         87           50 1               B             g2    </span>
<span class="co">#&gt;  4 TGAC… SeuratPro…        127           56 0               A             g2    </span>
<span class="co">#&gt;  5 AGTC… SeuratPro…        173           53 0               A             g2    </span>
<span class="co">#&gt;  6 TCTG… SeuratPro…         70           48 0               A             g1    </span>
<span class="co">#&gt;  7 TGGT… SeuratPro…         64           36 0               A             g1    </span>
<span class="co">#&gt;  8 GCAG… SeuratPro…         72           45 0               A             g1    </span>
<span class="co">#&gt;  9 GATA… SeuratPro…         52           36 0               A             g1    </span>
<span class="co">#&gt; 10 AATG… SeuratPro…        100           41 0               A             g1    </span>
<span class="co">#&gt; # … with 70 more rows, and 9 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span>
<span class="co">#&gt; #   PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   tSNE_2 &lt;dbl&gt;</span></pre></div>
<p><strong>But it is a Seurat object after all</strong></p>
<div class="sourceCode" id="cb58"><pre class="downlit">
<span class="va">pbmc_small_tidy</span><span class="op">@</span><span class="va">assays</span>
<span class="co">#&gt; $RNA</span>
<span class="co">#&gt; Assay data with 230 features for 80 cells</span>
<span class="co">#&gt; Top 10 variable features:</span>
<span class="co">#&gt;  PPBP, IGLL5, VDAC3, CD1C, AKR1C3, PF4, MYL9, GNLY, TREML1, CA2</span></pre></div>
</div>
<div id="preprocess-the-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#preprocess-the-dataset" class="anchor"></a>Preprocess the dataset</h2>
<div class="sourceCode" id="cb59"><pre class="downlit">
<span class="va">pbmc_small_pca</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/SCTransform.html">SCTransform</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">pbmc_small_pca</span>
<span class="co">#&gt; # A tibble: 80 x 18</span>
<span class="co">#&gt;    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span>
<span class="co">#&gt;  1 ATGC… SeuratPro…         70           47 0               A             g2    </span>
<span class="co">#&gt;  2 CATG… SeuratPro…         85           52 0               A             g1    </span>
<span class="co">#&gt;  3 GAAC… SeuratPro…         87           50 1               B             g2    </span>
<span class="co">#&gt;  4 TGAC… SeuratPro…        127           56 0               A             g2    </span>
<span class="co">#&gt;  5 AGTC… SeuratPro…        173           53 0               A             g2    </span>
<span class="co">#&gt;  6 TCTG… SeuratPro…         70           48 0               A             g1    </span>
<span class="co">#&gt;  7 TGGT… SeuratPro…         64           36 0               A             g1    </span>
<span class="co">#&gt;  8 GCAG… SeuratPro…         72           45 0               A             g1    </span>
<span class="co">#&gt;  9 GATA… SeuratPro…         52           36 0               A             g1    </span>
<span class="co">#&gt; 10 AATG… SeuratPro…        100           41 0               A             g1    </span>
<span class="co">#&gt; # … with 70 more rows, and 11 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span>
<span class="co">#&gt; #   nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></pre></div>
</div>
<div id="identify-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#identify-clusters" class="anchor"></a>Identify clusters</h2>
<p>We proceed with cluster identification with Seurat.</p>
<div class="sourceCode" id="cb60"><pre class="downlit">
<span class="va">pbmc_small_cluster</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_pca</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"igraph"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">pbmc_small_cluster</span>
<span class="co">#&gt; # A tibble: 80 x 20</span>
<span class="co">#&gt;    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span>
<span class="co">#&gt;  1 ATGC… SeuratPro…         70           47 0               A             g2    </span>
<span class="co">#&gt;  2 CATG… SeuratPro…         85           52 0               A             g1    </span>
<span class="co">#&gt;  3 GAAC… SeuratPro…         87           50 1               B             g2    </span>
<span class="co">#&gt;  4 TGAC… SeuratPro…        127           56 0               A             g2    </span>
<span class="co">#&gt;  5 AGTC… SeuratPro…        173           53 0               A             g2    </span>
<span class="co">#&gt;  6 TCTG… SeuratPro…         70           48 0               A             g1    </span>
<span class="co">#&gt;  7 TGGT… SeuratPro…         64           36 0               A             g1    </span>
<span class="co">#&gt;  8 GCAG… SeuratPro…         72           45 0               A             g1    </span>
<span class="co">#&gt;  9 GATA… SeuratPro…         52           36 0               A             g1    </span>
<span class="co">#&gt; 10 AATG… SeuratPro…        100           41 0               A             g1    </span>
<span class="co">#&gt; # … with 70 more rows, and 13 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span>
<span class="co">#&gt; #   nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, SCT_snn_res.0.8 &lt;fct&gt;,</span>
<span class="co">#&gt; #   seurat_clusters &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></pre></div>
<p>Now we can interrogate the object as if it was a regular tibble data frame.</p>
<div class="sourceCode" id="cb61"><pre class="downlit">
<span class="va">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu">tidyseurat</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/count.html">count</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">seurat_clusters</span><span class="op">)</span>
<span class="co">#&gt; tidyseurat says: A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 8 x 3</span>
<span class="co">#&gt;   groups seurat_clusters     n</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;fct&gt;           &lt;int&gt;</span>
<span class="co">#&gt; 1 g1     0                  17</span>
<span class="co">#&gt; 2 g1     1                  14</span>
<span class="co">#&gt; 3 g1     2                   9</span>
<span class="co">#&gt; 4 g1     3                   4</span>
<span class="co">#&gt; 5 g2     0                  13</span>
<span class="co">#&gt; 6 g2     1                  12</span>
<span class="co">#&gt; 7 g2     2                   6</span>
<span class="co">#&gt; 8 g2     3                   5</span></pre></div>
<p>We can identify cluster markers using Seurat.</p>
<div class="sourceCode" id="cb62"><pre class="downlit">
<span class="co"># Identify top 10 markers per cluster</span>
<span class="va">markers</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span>only.pos <span class="op">=</span> <span class="cn">TRUE</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span>, thresh.use <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span><span class="fl">10</span>, <span class="va">avg_logFC</span><span class="op">)</span>
<span class="co">#&gt; Calculating cluster 0</span>
<span class="co">#&gt; Calculating cluster 1</span>
<span class="co">#&gt; Calculating cluster 2</span>
<span class="co">#&gt; Calculating cluster 3</span>

<span class="co"># Plot heatmap</span>
<span class="va">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DoHeatmap.html">DoHeatmap</a></span><span class="op">(</span>
    features <span class="op">=</span> <span class="va">markers</span><span class="op">$</span><span class="va">gene</span>,
    group.colors <span class="op">=</span> <span class="va">friendly_cols</span>
  <span class="op">)</span></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-63-1.png" width="700"></p>
</div>
<div id="reduce-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce dimensions</h2>
<p>We can calculate the first 3 UMAP dimensions using the Seurat framework.</p>
<div class="sourceCode" id="cb63"><pre class="downlit">
<span class="va">pbmc_small_UMAP</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, n.components <span class="op">=</span> <span class="fl">3L</span>, <span class="op">)</span>
<span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span>
<span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span>
<span class="co">#&gt; This message will be shown once per session</span>
<span class="co">#&gt; 03:50:43 UMAP embedding parameters a = 0.9922 b = 1.112</span>
<span class="co">#&gt; 03:50:43 Read 80 rows and found 15 numeric columns</span>
<span class="co">#&gt; 03:50:43 Using Annoy for neighbor search, n_neighbors = 30</span>
<span class="co">#&gt; 03:50:43 Building Annoy index with metric = cosine, n_trees = 50</span>
<span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span>
<span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span>
<span class="co">#&gt; **************************************************|</span>
<span class="co">#&gt; 03:50:43 Writing NN index file to temp file /tmp/RtmpgoeXJX/file206e44c548c2</span>
<span class="co">#&gt; 03:50:43 Searching Annoy index using 1 thread, search_k = 3000</span>
<span class="co">#&gt; 03:50:43 Annoy recall = 100%</span>
<span class="co">#&gt; 03:50:44 Commencing smooth kNN distance calibration using 1 thread</span>
<span class="co">#&gt; 03:50:46 Initializing from normalized Laplacian + noise</span>
<span class="co">#&gt; 03:50:46 Commencing optimization for 500 epochs, with 2040 positive edges</span>
<span class="co">#&gt; 03:50:47 Optimization finished</span></pre></div>
<p>And we can plot them using 3D plot using plotly.</p>
<div class="sourceCode" id="cb64"><pre class="downlit">
<span class="va">pbmc_small_UMAP</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/plot_ly.html">plot_ly</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP_1`</span>,
    y <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP_2`</span>,
    z <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP_3`</span>,
    color <span class="op">=</span> <span class="op">~</span><span class="va">seurat_clusters</span>,
    colors <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>
  <span class="op">)</span></pre></div>
<p><img src="../inst/vignettes/plotly.png" width="340"></p>
</div>
<div id="cell-type-classification" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-classification" class="anchor"></a>Cell type classification</h2>
<p>We can infer cell type identities using <em>SingleR</em> <span class="citation">(Aran et al. 2019)</span> and manipulate the output using tidyverse.</p>
<div class="sourceCode" id="cb65"><pre class="downlit">
<span class="co"># Get cell type reference data</span>
<span class="va">blueprint</span> <span class="op">&lt;-</span> <span class="fu">celldex</span><span class="fu">::</span><span class="fu">BlueprintEncodeData</span><span class="op">(</span><span class="op">)</span>

<span class="co"># Infer cell identities</span>
<span class="va">cell_type_df</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_UMAP</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"SCT"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html">Matrix</a></span><span class="op">(</span>sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">SingleR</span><span class="fu">::</span><span class="fu">SingleR</span><span class="op">(</span>
    ref <span class="op">=</span> <span class="va">blueprint</span>,
    labels <span class="op">=</span> <span class="va">blueprint</span><span class="op">$</span><span class="va">label.main</span>,
    method <span class="op">=</span> <span class="st">"single"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/select.html">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">first.labels</span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb66"><pre class="downlit">
<span class="co"># Join UMAP and cell type info</span>
<span class="va">pbmc_small_cell_type</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_UMAP</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/left_join.html">left_join</a></span><span class="op">(</span><span class="va">cell_type_df</span>, by <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span>

<span class="co"># Reorder columns</span>
<span class="va">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu">tidyseurat</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/select.html">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">first.labels</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 80 x 24</span>
<span class="co">#&gt;    cell  first.labels orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;        &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;          </span>
<span class="co">#&gt;  1 ATGC… CD4+ T-cells SeuratPro…         70           47 0              </span>
<span class="co">#&gt;  2 CATG… CD8+ T-cells SeuratPro…         85           52 0              </span>
<span class="co">#&gt;  3 GAAC… CD8+ T-cells SeuratPro…         87           50 1              </span>
<span class="co">#&gt;  4 TGAC… CD4+ T-cells SeuratPro…        127           56 0              </span>
<span class="co">#&gt;  5 AGTC… CD4+ T-cells SeuratPro…        173           53 0              </span>
<span class="co">#&gt;  6 TCTG… CD4+ T-cells SeuratPro…         70           48 0              </span>
<span class="co">#&gt;  7 TGGT… CD4+ T-cells SeuratPro…         64           36 0              </span>
<span class="co">#&gt;  8 GCAG… CD4+ T-cells SeuratPro…         72           45 0              </span>
<span class="co">#&gt;  9 GATA… CD4+ T-cells SeuratPro…         52           36 0              </span>
<span class="co">#&gt; 10 AATG… CD4+ T-cells SeuratPro…        100           41 0              </span>
<span class="co">#&gt; # … with 70 more rows, and 18 more variables: letter.idents &lt;fct&gt;,</span>
<span class="co">#&gt; #   groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, nCount_SCT &lt;dbl&gt;,</span>
<span class="co">#&gt; #   nFeature_SCT &lt;int&gt;, SCT_snn_res.0.8 &lt;fct&gt;, seurat_clusters &lt;fct&gt;,</span>
<span class="co">#&gt; #   PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   tSNE_2 &lt;dbl&gt;, UMAP_1 &lt;dbl&gt;, UMAP_2 &lt;dbl&gt;, UMAP_3 &lt;dbl&gt;</span></pre></div>
<p>We can easily summarise the results. For example, we can see how cell type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb67"><pre class="downlit">
<span class="va">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/count.html">count</a></span><span class="op">(</span><span class="va">seurat_clusters</span>, <span class="va">first.labels</span><span class="op">)</span>
<span class="co">#&gt; tidyseurat says: A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 9 x 3</span>
<span class="co">#&gt;   seurat_clusters first.labels     n</span>
<span class="co">#&gt;   &lt;fct&gt;           &lt;chr&gt;        &lt;int&gt;</span>
<span class="co">#&gt; 1 0               CD4+ T-cells     8</span>
<span class="co">#&gt; 2 0               CD8+ T-cells    10</span>
<span class="co">#&gt; 3 0               NK cells        12</span>
<span class="co">#&gt; 4 1               Macrophages      1</span>
<span class="co">#&gt; 5 1               Monocytes       25</span>
<span class="co">#&gt; 6 2               B-cells         10</span>
<span class="co">#&gt; 7 2               Macrophages      1</span>
<span class="co">#&gt; 8 2               Monocytes        4</span>
<span class="co">#&gt; 9 3               Erythrocytes     9</span></pre></div>
</div>
<div id="nested-analyses-1" class="section level2">
<h2 class="hasAnchor">
<a href="#nested-analyses-1" class="anchor"></a>Nested analyses</h2>
<p>A powerful tool we can use with tidyseurat is <code>nest</code>. We can easily perform independent analyses on subsets of the dataset. First we classify cell types in lymphoid and myeloid; then, nest based on the new classification</p>
<div class="sourceCode" id="cb68"><pre class="downlit">
<span class="va">pbmc_small_nested</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/filter.html">filter</a></span><span class="op">(</span><span class="va">first.labels</span> <span class="op">!=</span> <span class="st">"Erythrocytes"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/mutate.html">mutate</a></span><span class="op">(</span>cell_class <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">`first.labels`</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Macrophages"</span>, <span class="st">"Monocytes"</span><span class="op">)</span>, <span class="st">"myeloid"</span>, <span class="st">"lymphoid"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/tidyr-methods.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">cell_class</span><span class="op">)</span>

<span class="va">pbmc_small_nested</span>
<span class="co">#&gt; # A tibble: 2 x 2</span>
<span class="co">#&gt;   cell_class data      </span>
<span class="co">#&gt;   &lt;chr&gt;      &lt;list&gt;    </span>
<span class="co">#&gt; 1 lymphoid   &lt;tidysert&gt;</span>
<span class="co">#&gt; 2 myeloid    &lt;tidysert&gt;</span></pre></div>
<p>Now we can independently for the lymphoid and myeloid subsets (i) find variable features, (ii) reduce dimensions, and (iii) cluster using both tidyverse and SingleCellExperiment seamlessly.</p>
<div class="sourceCode" id="cb69"><pre class="downlit">
<span class="va">pbmc_small_nested_reanalysed</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_small_nested</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/mutate.html">mutate</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span>
    <span class="va">data</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span><span class="op">(</span>npcs <span class="op">=</span> <span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"igraph"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, n.components <span class="op">=</span> <span class="fl">3L</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
  <span class="op">)</span><span class="op">)</span>

<span class="va">pbmc_small_nested_reanalysed</span>
<span class="co">#&gt; # A tibble: 2 x 2</span>
<span class="co">#&gt;   cell_class data      </span>
<span class="co">#&gt;   &lt;chr&gt;      &lt;list&gt;    </span>
<span class="co">#&gt; 1 lymphoid   &lt;tidysert&gt;</span>
<span class="co">#&gt; 2 myeloid    &lt;tidysert&gt;</span></pre></div>
</div>
<div id="key-points-2" class="section level2">
<h2 class="hasAnchor">
<a href="#key-points-2" class="anchor"></a>Key Points</h2>
<ul>
<li>Some basic steps of a single-cell RNA sequencing analysis are dimensionality reduction, cluster identification and cell type classification</li>
<li>
<code>tidyseurat</code> is an invisible layer that operates on a <code>Seurat</code> object and enables us to visualise and manipulate data as if it were a tidy data frame.</li>
<li>
<code>tidyseurat</code> object is a <code>Seurat object</code> so it can be used with any <code>Seurat</code> compatible method</li>
</ul>
</div>
</div>
<div id="contributing" class="section level1">
<h1 class="hasAnchor">
<a href="#contributing" class="anchor"></a>Contributing</h1>
<p>If you want to suggest improvements for this workshop or ask questions, you can do so as described <a href="https://github.com/stemangiola/rpharma2020_tidytranscriptomics/blob/master/CONTRIBUTING.md">here</a>.</p>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>Record package and version information with <code>sessionInfo</code></p>
<div class="sourceCode" id="cb70"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.2 Patched (2020-09-24 r79253)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.1 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] tidyseurat_0.1.8            Seurat_3.2.2               </span>
<span class="co">#&gt;  [3] purrr_0.3.4                 survival_3.2-7             </span>
<span class="co">#&gt;  [5] GGally_2.0.0                tidygate_0.2.8             </span>
<span class="co">#&gt;  [7] forcats_0.5.0               plotly_4.9.2.1             </span>
<span class="co">#&gt;  [9] ggrepel_0.8.2               tidybulk_1.1.8             </span>
<span class="co">#&gt; [11] tidyHeatmap_1.1.5           ggplot2_3.3.2              </span>
<span class="co">#&gt; [13] stringr_1.4.0               readr_1.3.1                </span>
<span class="co">#&gt; [15] tidyr_1.1.2                 dplyr_1.0.2                </span>
<span class="co">#&gt; [17] tibble_3.0.3                org.Dm.eg.db_3.11.4        </span>
<span class="co">#&gt; [19] org.Hs.eg.db_3.11.4         AnnotationDbi_1.51.3       </span>
<span class="co">#&gt; [21] airway_1.9.0                SummarizedExperiment_1.19.8</span>
<span class="co">#&gt; [23] Biobase_2.49.1              GenomicRanges_1.41.6       </span>
<span class="co">#&gt; [25] GenomeInfoDb_1.25.11        IRanges_2.23.10            </span>
<span class="co">#&gt; [27] S4Vectors_0.27.13           BiocGenerics_0.35.4        </span>
<span class="co">#&gt; [29] MatrixGenerics_1.1.3        matrixStats_0.57.0         </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] utf8_1.1.4                           reticulate_1.16                     </span>
<span class="co">#&gt;   [3] tidyselect_1.1.0                     RSQLite_2.2.1                       </span>
<span class="co">#&gt;   [5] htmlwidgets_1.5.1                    grid_4.0.2                          </span>
<span class="co">#&gt;   [7] BiocParallel_1.23.2                  Rtsne_0.15                          </span>
<span class="co">#&gt;   [9] munsell_0.5.0                        codetools_0.2-16                    </span>
<span class="co">#&gt;  [11] ragg_0.3.1                           ica_1.0-2                           </span>
<span class="co">#&gt;  [13] preprocessCore_1.51.0                future_1.19.1                       </span>
<span class="co">#&gt;  [15] miniUI_0.1.1.1                       withr_2.3.0                         </span>
<span class="co">#&gt;  [17] colorspace_1.4-1                     knitr_1.30                          </span>
<span class="co">#&gt;  [19] SingleCellExperiment_1.11.7          ROCR_1.0-11                         </span>
<span class="co">#&gt;  [21] tensor_1.5                           rpharma2020tidytranscriptomics_1.0.6</span>
<span class="co">#&gt;  [23] listenv_0.8.0                        labeling_0.3                        </span>
<span class="co">#&gt;  [25] GenomeInfoDbData_1.2.3               polyclip_1.10-0                     </span>
<span class="co">#&gt;  [27] bit64_4.0.5                          farver_2.0.3                        </span>
<span class="co">#&gt;  [29] pheatmap_1.0.12                      rprojroot_1.3-2                     </span>
<span class="co">#&gt;  [31] vctrs_0.3.4                          generics_0.0.2                      </span>
<span class="co">#&gt;  [33] xfun_0.18                            dittoSeq_1.1.9                      </span>
<span class="co">#&gt;  [35] R6_2.4.1                             clue_0.3-57                         </span>
<span class="co">#&gt;  [37] rsvd_1.0.3                           locfit_1.5-9.4                      </span>
<span class="co">#&gt;  [39] bitops_1.0-6                         spatstat.utils_1.17-0               </span>
<span class="co">#&gt;  [41] reshape_0.8.8                        DelayedArray_0.15.12                </span>
<span class="co">#&gt;  [43] assertthat_0.2.1                     promises_1.1.1                      </span>
<span class="co">#&gt;  [45] scales_1.1.1                         gtable_0.3.0                        </span>
<span class="co">#&gt;  [47] globals_0.13.0                       goftest_1.2-2                       </span>
<span class="co">#&gt;  [49] rlang_0.4.7                          genefilter_1.71.0                   </span>
<span class="co">#&gt;  [51] systemfonts_0.3.2                    GlobalOptions_0.1.2                 </span>
<span class="co">#&gt;  [53] splines_4.0.2                        lazyeval_0.2.2                      </span>
<span class="co">#&gt;  [55] broom_0.7.1                          yaml_2.2.1                          </span>
<span class="co">#&gt;  [57] reshape2_1.4.4                       abind_1.4-5                         </span>
<span class="co">#&gt;  [59] crosstalk_1.1.0.1                    backports_1.1.10                    </span>
<span class="co">#&gt;  [61] httpuv_1.5.4                         tools_4.0.2                         </span>
<span class="co">#&gt;  [63] ellipsis_0.3.1                       RColorBrewer_1.1-2                  </span>
<span class="co">#&gt;  [65] ggridges_0.5.2                       Rcpp_1.0.5                          </span>
<span class="co">#&gt;  [67] plyr_1.8.6                           zlibbioc_1.35.0                     </span>
<span class="co">#&gt;  [69] RCurl_1.98-1.2                       rpart_4.1-15                        </span>
<span class="co">#&gt;  [71] deldir_0.1-29                        pbapply_1.4-3                       </span>
<span class="co">#&gt;  [73] GetoptLong_1.0.3                     viridis_0.5.1                       </span>
<span class="co">#&gt;  [75] cowplot_1.1.0                        zoo_1.8-8                           </span>
<span class="co">#&gt;  [77] cluster_2.1.0                        fs_1.5.0                            </span>
<span class="co">#&gt;  [79] magrittr_1.5                         RSpectra_0.16-0                     </span>
<span class="co">#&gt;  [81] data.table_1.13.0                    circlize_0.4.10                     </span>
<span class="co">#&gt;  [83] lmtest_0.9-38                        RANN_2.6.1                          </span>
<span class="co">#&gt;  [85] fitdistrplus_1.1-1                   hms_0.5.3                           </span>
<span class="co">#&gt;  [87] patchwork_1.0.1                      mime_0.9                            </span>
<span class="co">#&gt;  [89] evaluate_0.14                        xtable_1.8-4                        </span>
<span class="co">#&gt;  [91] XML_3.99-0.5                         gridExtra_2.3                       </span>
<span class="co">#&gt;  [93] shape_1.4.5                          compiler_4.0.2                      </span>
<span class="co">#&gt;  [95] KernSmooth_2.23-17                   crayon_1.3.4                        </span>
<span class="co">#&gt;  [97] htmltools_0.5.0                      mgcv_1.8-33                         </span>
<span class="co">#&gt;  [99] later_1.1.0.1                        geneplotter_1.67.0                  </span>
<span class="co">#&gt; [101] DBI_1.1.0                            ComplexHeatmap_2.5.5                </span>
<span class="co">#&gt; [103] MASS_7.3-53                          rappdirs_0.3.1                      </span>
<span class="co">#&gt; [105] boot_1.3-25                          Matrix_1.2-18                       </span>
<span class="co">#&gt; [107] cli_2.0.2                            igraph_1.2.5                        </span>
<span class="co">#&gt; [109] pkgconfig_2.0.3                      pkgdown_1.6.1                       </span>
<span class="co">#&gt; [111] annotate_1.67.1                      XVector_0.29.3                      </span>
<span class="co">#&gt; [113] digest_0.6.25                        sctransform_0.3                     </span>
<span class="co">#&gt; [115] RcppAnnoy_0.0.16                     spatstat.data_1.4-3                 </span>
<span class="co">#&gt; [117] rmarkdown_2.4                        leiden_0.3.3                        </span>
<span class="co">#&gt; [119] uwot_0.1.8                           edgeR_3.31.4                        </span>
<span class="co">#&gt; [121] shiny_1.5.0                          rjson_0.2.20                        </span>
<span class="co">#&gt; [123] lifecycle_0.2.0                      nlme_3.1-149                        </span>
<span class="co">#&gt; [125] jsonlite_1.7.1                       desc_1.2.0                          </span>
<span class="co">#&gt; [127] viridisLite_0.3.0                    limma_3.45.14                       </span>
<span class="co">#&gt; [129] fansi_0.4.1                          pillar_1.4.6                        </span>
<span class="co">#&gt; [131] lattice_0.20-41                      fastmap_1.0.1                       </span>
<span class="co">#&gt; [133] httr_1.4.2                           glue_1.4.2                          </span>
<span class="co">#&gt; [135] spatstat_1.64-1                      png_0.1-7                           </span>
<span class="co">#&gt; [137] bit_4.0.4                            stringi_1.5.3                       </span>
<span class="co">#&gt; [139] blob_1.2.1                           DESeq2_1.29.14                      </span>
<span class="co">#&gt; [141] memoise_1.1.0                        irlba_2.3.3                         </span>
<span class="co">#&gt; [143] future.apply_1.6.0</span></pre></div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-aran2019reference">
<p>Aran, Dvir, Agnieszka P Looney, Leqian Liu, Esther Wu, Valerie Fong, Austin Hsu, Suzanna Chak, et al. 2019. “Reference-Based Analysis of Lung Single-Cell Sequencing Reveals a Transitional Profibrotic Macrophage.” <em>Nature Immunology</em> 20 (2): 163–72.</p>
</div>
<div id="ref-brooks2011conservation">
<p>Brooks, Angela N, Li Yang, Michael O Duff, Kasper D Hansen, Jung W Park, Sandrine Dudoit, Steven E Brenner, and Brenton R Graveley. 2011. “Conservation of an Rna Regulatory Map Between Drosophila and Mammals.” <em>Genome Research</em> 21 (2): 193–202.</p>
</div>
<div id="ref-butler2018integrating">
<p>Butler, Andrew, Paul Hoffman, Peter Smibert, Efthymia Papalexi, and Rahul Satija. 2018. “Integrating Single-Cell Transcriptomic Data Across Different Conditions, Technologies, and Species.” <em>Nature Biotechnology</em> 36 (5): 411–20.</p>
</div>
<div id="ref-chen2016reads">
<p>Chen, Yunshun, Aaron TL Lun, and Gordon K Smyth. 2016. “From Reads to Genes to Pathways: Differential Expression Analysis of Rna-Seq Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.” <em>F1000Research</em> 5.</p>
</div>
<div id="ref-himes2014rna">
<p>Himes, Blanca E, Xiaofeng Jiang, Peter Wagner, Ruoxi Hu, Qiyu Wang, Barbara Klanderman, Reid M Whitaker, et al. 2014. “RNA-Seq Transcriptome Profiling Identifies Crispld2 as a Glucocorticoid Responsive Gene That Modulates Cytokine Function in Airway Smooth Muscle Cells.” <em>PloS One</em> 9 (6): e99625.</p>
</div>
<div id="ref-law2016rna">
<p>Law, Charity W, Monther Alhamdoosh, Shian Su, Xueyi Dong, Luyi Tian, Gordon K Smyth, and Matthew E Ritchie. 2016. “RNA-Seq Analysis Is Easy as 1-2-3 with Limma, Glimma and edgeR.” <em>F1000Research</em> 5.</p>
</div>
<div id="ref-law2014voom">
<p>Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. “Voom: Precision Weights Unlock Linear Model Analysis Tools for Rna-Seq Read Counts.” <em>Genome Biology</em> 15 (2): R29.</p>
</div>
<div id="ref-love2014moderated">
<p>Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for Rna-Seq Data with Deseq2.” <em>Genome Biology</em> 15 (12): 550.</p>
</div>
<div id="ref-mangiola2020tidyheatmap">
<p>Mangiola, Stefano, and Anthony T Papenfuss. 2020. “TidyHeatmap: An R Package for Modular Heatmap Production Based on Tidy Principles.” <em>Journal of Open Source Software</em> 5 (52): 2472.</p>
</div>
<div id="ref-mccarthy2012differential">
<p>McCarthy, Davis J, Yunshun Chen, and Gordon K Smyth. 2012. “Differential Expression Analysis of Multifactor Rna-Seq Experiments with Respect to Biological Variation.” <em>Nucleic Acids Research</em> 40 (10): 4288–97.</p>
</div>
<div id="ref-robinson2010edger">
<p>Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010. “EdgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics</em> 26 (1): 139–40.</p>
</div>
<div id="ref-robinson2010scaling">
<p>Robinson, Mark D, and Alicia Oshlack. 2010. “A Scaling Normalization Method for Differential Expression Analysis of Rna-Seq Data.” <em>Genome Biology</em> 11 (3): 1–9.</p>
</div>
<div id="ref-stuart2019comprehensive">
<p>Stuart, Tim, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck III, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. 2019. “Comprehensive Integration of Single-Cell Data.” <em>Cell</em> 177 (7): 1888–1902.</p>
</div>
<div id="ref-wickham2019welcome">
<p>Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. “Welcome to the Tidyverse.” <em>Journal of Open Source Software</em> 4 (43): 1686.</p>
</div>
<div id="ref-wickham2014tidy">
<p>Wickham, Hadley, and others. 2014. “Tidy Data.” <em>Journal of Statistical Software</em> 59 (10): 1–23.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>&lt;maria.doyle at petermac.org&gt;<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>&lt;mangiola.s at wehi.edu.au&gt;<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Maria Doyle, Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
